<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Selection of relevant environment and genotype variables associated with Variably Methylated Loci (VML) — selectVariables • RAMEN</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Selection of relevant environment and genotype variables associated with Variably Methylated Loci (VML) — selectVariables"><meta name="description" content="For each VML, this function selects potentially relevant genotype and environmental variables associated with DNA methylation levels of said VML using LASSO. See details below for more information."><meta property="og:description" content="For each VML, this function selects potentially relevant genotype and environmental variables associated with DNA methylation levels of said VML using LASSO. See details below for more information."><meta property="og:image" content="https://ericknavarrod.github.io/RAMEN/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">RAMEN</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/RAMEN.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ErickNavarroD/RAMEN/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Selection of relevant environment and genotype variables associated with Variably Methylated Loci (VML)</h1>
      <small class="dont-index">Source: <a href="https://github.com/ErickNavarroD/RAMEN/blob/main/R/selectVariables.R" class="external-link"><code>R/selectVariables.R</code></a></small>
      <div class="d-none name"><code>selectVariables.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>For each VML, this function selects potentially relevant genotype and environmental variables associated with DNA methylation levels of said VML using LASSO. See details below for more information.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">selectVariables</span><span class="op">(</span></span>
<span>  <span class="va">VML_df</span>,</span>
<span>  <span class="va">genotype_matrix</span>,</span>
<span>  <span class="va">environmental_matrix</span>,</span>
<span>  covariates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">summarized_methyl_VML</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-vml-df">VML_df<a class="anchor" aria-label="anchor" href="#arg-vml-df"></a></dt>
<dd><p>A data frame converted from a GRanges object. Recommended to use the output of <em>RAMEN::findCisSNPs()</em>. Must have one VML per row, and contain the following columns: "VML_index" (a unique ID for each VML in VML_df AS CHARACTERS) and "SNP" (a column with a list as observation, containing the name of the SNPs surrounding the corresponding VML).  The SNPs contained in the "SNP" column must be present in the object that is indicated in the genotype_matrix argument, and it must contain all the VML contained in summarized_methyl_VML. VML with no surrounding SNPs must have an empty list in the SNP column (either list(NULL), list(NA), list("") or list(character(0)) ).</p></dd>


<dt id="arg-genotype-matrix">genotype_matrix<a class="anchor" aria-label="anchor" href="#arg-genotype-matrix"></a></dt>
<dd><p>A matrix of number-encoded genotypes. Columns must correspond to samples, and rows to SNPs. We suggest using a gene-dosage model, which would encode the SNPs ordinally depending on the genotype allele charge, such as 2 (AA), 1 (AB) and 0 (BB). The column names must correspond with individual IDs.</p></dd>


<dt id="arg-environmental-matrix">environmental_matrix<a class="anchor" aria-label="anchor" href="#arg-environmental-matrix"></a></dt>
<dd><p>A matrix of environmental variables. Only numeric values are supported. In case of factor variables, it is recommended to encode them as numbers or re-code them into dummy variables if there are more than two levels. Columns must correspond to environmental variables and rows to individuals. Row names must be the individual IDs.</p></dd>


<dt id="arg-covariates">covariates<a class="anchor" aria-label="anchor" href="#arg-covariates"></a></dt>
<dd><p>A matrix containing the covariates (i.e., concomitant variables / variables that are not the ones you are interested in) that will be adjusted for in the final GxE models (e.g., cell type proportions, age, etc.). Each column should correspond to a covariate and each row to an individual. Row names must correspond to the individual IDs.</p></dd>


<dt id="arg-summarized-methyl-vml">summarized_methyl_VML<a class="anchor" aria-label="anchor" href="#arg-summarized-methyl-vml"></a></dt>
<dd><p>A data frame containing each individual's VML summarized methylation. It is suggested to use the output of RAMEN::summarizeVML().Rows must reflects individuals, and columns VML The names of the columns must correspond to the index of said VML, and it must match the index of VML_df$VML_index. The names of the rows must correspond to the sample IDs, and must match with the IDs of the other matrices.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>An integer number that initializes a pseudo-random number generator. Random numbers in this function are created during the lambda cross validation and the LASSO stages. Setting a seed is highly encouraged for result reproducibility. <strong>Please note that setting a seed in this function modifies the seed globally</strong>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with three columns:</p><ul><li><p>VML_index: Unique VML ID.</p></li>
<li><p>selected_genot: Column containing lists as values with the selected SNPs.</p></li>
<li><p>selected_env: Column containing lists as values with the selected environmental variables.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>selectVariables() uses LASSO, which is an embedded variable selection method that penalizes models that are more complex (i.e., that contain more variables) in favor of simpler models (i.e. that contain less variables), but not at the expense of reducing predictive power. Using LASSO's variable screening property (with high probability, the LASSO estimated model includes the substantial covariates and drops the redundant ones) this function selects genotype and environment variables with potential relevance in the Variable Methylated Loci (VML) dataset (see also Bühlmann and van de Geer, 2011). For each VML, LASSO is run three times: 1) including only the genotype variables for the selection step, 2) including only the environmental variables for the selection step, and 3) Including both the genotype and environmental variables in the selection step. This is done to ensure that the function captures the variables that are relevant within their own category (e.g., SNPs that are strongly associated with the DNAme levels of a VML in the presence of the rest of the SNPs) or in the presence of the variables of the other category (e.g. SNPs that are strongly associated with the DNAme levels of a VML in the presence of the rest of BOTH the SNPs AND environmental variables). Every time LASSO is run, the basal covariates (i.e., concomitant variables )indicated in the argument <em>covariates</em> are not penalized (i.e., those variables are always included in the models and their coefficients are not subjected to shrinkage). That way, only the most promising E and G variables in the presence of the concomitant variables will be selected.</p>
<p>Each LASSO model uses a tuned lambda that minimizes the 5-fold cross-validation error within its corresponding data. This function uses the lambda.min value in contrast to lambda.1se because its goal within the RAMEN package is to use LASSO to reduce the number of variables that are going to be used next for fitting pairwise interaction models in <em>lmGE()</em>. Since at this step variables are being selected based only on main effects, it is preferable to cast a "wider net" and select a slightly higher number of variables that could potentially have a strong interaction effect when paired with another variable. Furthermore, since in this case LASSO is being used as a screening procedure to select variables that will be fit separately in independent models and compared, the overfitting issue of using lambda.min does not impose a big concern. After finding the best lambda value, the sequence of models is fit by coordinate descent using <em>glmnet()</em>. Random numbers in this function are created during the lambda cross validation and the LASSO stages. Setting a seed is highly encouraged for result reproducibility using the <em>seed</em> argument. Please note that setting a seed inside of this function modifies the seed globally (which is R's default behavior).</p>
<p>#' This function supports parallel computing for increased speed. To do so, you have to set the parallel back-end
in your R session before running the function  (e.g., <em>doParallel::registerDoParallel(4)</em>). After that, the function can be run as usual. It is recommended to also set options(future.globals.maxSize= +Inf). Please make sure that your data has no NAs and it's all numerical, since the LASSO implementation we use does not support missing or non-numerical values.</p>
<p>Note: If you want to conduct the variable selection step only in one data set (i.e., only in the genotype), you can set the argument <em>environmental_matrix = NULL</em>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Find VML in test data</span></span></span>
<span class="r-in"><span><span class="va">VML</span> <span class="op">&lt;-</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="findVML.html">findVML</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>   methylation_data <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="test_methylation_data.html">test_methylation_data</a></span>,</span></span>
<span class="r-in"><span>   array_manifest <span class="op">=</span> <span class="st">"IlluminaHumanMethylationEPICv1"</span>,</span></span>
<span class="r-in"><span>   cor_threshold <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>   var_method <span class="op">=</span> <span class="st">"variance"</span>,</span></span>
<span class="r-in"><span>   var_distribution <span class="op">=</span> <span class="st">"ultrastable"</span>,</span></span>
<span class="r-in"><span>   var_threshold_percentile <span class="op">=</span> <span class="fl">0.99</span>,</span></span>
<span class="r-in"><span>   max_distance <span class="op">=</span> <span class="fl">1000</span></span></span>
<span class="r-in"><span>   <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identifying Highly Variable Probes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identifying sparse Variable Methylated Probes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identifying Variable Methylated Regions...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Applying correlation filter to Variable Methylated Regions...</span>
<span class="r-in"><span><span class="co">## Find cis SNPs around VML</span></span></span>
<span class="r-in"><span><span class="va">VML_with_cis_snps</span> <span class="op">&lt;-</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="findCisSNPs.html">findCisSNPs</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  VML_df <span class="op">=</span> <span class="va">VML</span><span class="op">$</span><span class="va">VML</span>,</span></span>
<span class="r-in"><span>  genotype_information <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="test_genotype_information.html">test_genotype_information</a></span>,</span></span>
<span class="r-in"><span>  distance <span class="op">=</span> <span class="fl">1e6</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reminder: please make sure that the positions of the VML data frame and the ones in the genotype information are from the same genome build.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Summarize methylation levels in VML</span></span></span>
<span class="r-in"><span><span class="va">summarized_methyl_VML</span> <span class="op">&lt;-</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="summarizeVML.html">summarizeVML</a></span><span class="op">(</span></span></span>
<span class="r-in"><span> methylation_data <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="test_methylation_data.html">test_methylation_data</a></span>,</span></span>
<span class="r-in"><span> VML_df <span class="op">=</span> <span class="va">VML_with_cis_snps</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="co">## Select relevant genotype and environmental variables</span></span></span>
<span class="r-in"><span> <span class="va">selected_vars</span> <span class="op">&lt;-</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu">selectVariables</span><span class="op">(</span></span></span>
<span class="r-in"><span>   VML_df <span class="op">=</span> <span class="va">VML_with_cis_snps</span>,</span></span>
<span class="r-in"><span>   genotype_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="test_genotype_matrix.html">test_genotype_matrix</a></span>,</span></span>
<span class="r-in"><span>   environmental_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="test_environmental_matrix.html">test_environmental_matrix</a></span>,</span></span>
<span class="r-in"><span>   covariates <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="test_covariates.html">test_covariates</a></span>,</span></span>
<span class="r-in"><span>   summarized_methyl_VML <span class="op">=</span> <span class="va">summarized_methyl_VML</span>,</span></span>
<span class="r-in"><span>   seed <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Erick I. Navarro-Delgado.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

