<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="RAMEN">
<title>RAMEN • RAMEN</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="RAMEN">
<meta property="og:description" content="RAMEN">
<meta property="og:image" content="https://ericknavarrod.github.io/RAMEN/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">RAMEN</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/RAMEN.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>RAMEN</h1>
                        <h4 data-toc-skip class="author">Erick I.
Navarro-Delgado</h4>
            <address class="author_afil">
      The University of British
Columbia<br><a class="author_email" href="mailto:#"></a><a href="mailto:erick.navarrodelgado@bcchr.ca" class="email">erick.navarrodelgado@bcchr.ca</a>
      </address>
                  
      
      
      <div class="d-none name"><code>RAMEN.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>Regional Association of Methylation variability with genotype
and ENvironment (RAMEN)</strong> is an R package whose goal is to
integrate genomic, methylomic and exposomic data to model the
contribution of genetics (G) and the environment (E) to DNA methylation
(DNAme) variability. RAMEN identifies Variable Methylated Regions (VMRs)
in microarray DNAme data and then, using genotype and environmental
data, it identifies which of the following models better explains this
variability in regions across the methylome:</p>
<table class="table">
<caption>Fitted models</caption>
<thead><tr class="header">
<th align="left">Model</th>
<th align="left">Name</th>
<th align="left">Abbreviation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">DNAme ~ G + covars</td>
<td align="left">Genetics</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="left">DNAme ~ E + covars</td>
<td align="left">Environmental exposure</td>
<td align="left">E</td>
</tr>
<tr class="odd">
<td align="left">DNAme ~ G + E + covars</td>
<td align="left">Additive</td>
<td align="left">G+E</td>
</tr>
<tr class="even">
<td align="left">DNAme ~ G + E + G*E + covars</td>
<td align="left">Interaction</td>
<td align="left">GxE</td>
</tr>
</tbody>
</table>
<p>where G variables are represented by SNPs, E variables by
environmental exposures, and where covars are concomitant variables
(i.e. variables that are adjusted for in the model and not of interest
in the study such as cell type proportion, age, etc.).</p>
<p>The main <a href="#gene-environment-interaction-analysis">gene-environment
interaction modeling</a> pipeline is conducted though six core
functions:</p>
<ul>
<li>
<code><a href="../reference/findVMRs.html">findVMRs()</a></code> identifies Variable Methylated Regions
(VMRs) in microarrays</li>
<li>
<code><a href="../reference/summarizeVMRs.html">summarizeVMRs()</a></code>summarizes the regional methylation
state of each VMR</li>
<li>
<code><a href="../reference/findCisSNPs.html">findCisSNPs()</a></code> identifies the SNPs in <em>cis</em> of
each VMR</li>
<li>
<code><a href="../reference/selectVariables.html">selectVariables()</a></code> conducts a LASSO-based variable
selection strategy to identify potentially relevant <em>cis</em> SNPs
and environmental variables</li>
<li>
<code><a href="../reference/lmGE.html">lmGE()</a></code> fits linear single-variable genetic (G) and
environmental (E), and pairwise additive (G+E) and interaction (GxE)
linear models and select the best explanatory model per VMR.</li>
<li>
<code><a href="../reference/nullDistGE.html">nullDistGE()</a></code> simulates a delta R squared null
distribution of G and E effects on DNAme variability. Useful for
filtering out poor-performing best explanatory models selected by
<em>lmGE()</em>.</li>
</ul>
<p>These functions are compatible with parallel computing, which is
recommended due to the computationally intensive tasks conducted by the
package.</p>
<p>In addition to the <a href="#gene-environment-interaction-analysis">standard gene-environment
interaction modeling pipeline</a>, RAMEN can be useful for other DNAme
analyses (see [ Variations to the standard workflow][]), such as
reducing the tested sites in Epigenome Wide Association Studies,
grouping DNAme probes into regions, identifying SNPs near a probe,
etc.</p>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>The manuscript detailing RAMEN and its use is currently under
preparation. For more information about this please contact Erick I.
Navarro-Delgado at <a href="mailto:erick.navarrodelgado@bcchr.ca" class="email">erick.navarrodelgado@bcchr.ca</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="gene-environment-interaction-analysis">Gene-environment interaction analysis<a class="anchor" aria-label="anchor" href="#gene-environment-interaction-analysis"></a>
</h2>
<p>The main purpose of the RAMEN package is to conduct a methylome-wide
analysis to identify which model (G, E, G+E or GxE) better explains the
variability across the genome. In this vignette, we will illustrate how
to use the package.</p>
<p>To conduct this analysis, the following cleaned data sets (i.e. after
quality and exploratory data analysis checks) from a cohort are
required:</p>
<ul>
<li>DNAme data</li>
<li>DNAme array manifest</li>
<li>Genotyping data</li>
<li>Genotype information</li>
<li>Environmental exposure data</li>
<li>Concomitant variables data</li>
</ul>
<p>Once that we have that data, the overview of the pipeline is the
following:</p>
<div class="figure">
<img src="RAMEN_pipeline.png" alt=""><p class="caption">RAMEN pipeline</p>
</div>
<p>where:</p>
<ul>
<li>DNAme data is grouped into VMRs, and then the DNAme state per
individual is summarized in each region</li>
<li>Using the identified VMRs and the genomic information, we identify
the SNPs in <em>cis</em> for each VMR</li>
<li>Both the <em>cis</em> SNPs and the exposome data are subjected to
the variable selection stage</li>
<li>The selected variables (SNPs a and Es) enter the modelling stage,
which outputs one single winning model per VMR</li>
<li>The thresholds obtained from the simulated null distribution are
used to remove winning models which performance are likely to be due to
chance.</li>
</ul>
<p>In the following sections we will go through each of these steps and
guide the user regarding the recommended parameters to use in each
function of the package. For illustration purposes, we provide small toy
data sets that do not intend to simulate a real biological phenomenon.
These data sets are already available in the RAMEN package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Load the packages used throughout the vignette</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ericknavarrod.github.io/RAMEN/">RAMEN</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="identify-vmrs-and-summarize-their-methylation-state">Identify VMRs and summarize their methylation state<a class="anchor" aria-label="anchor" href="#identify-vmrs-and-summarize-their-methylation-state"></a>
</h3>
<p>The first step of the pipeline is to identify the <strong>Variable
Methylated Regions</strong>(VMRs) in the data set. You might be
wondering <em>“What is a VMR and why do we use them instead of DNAme
levels from each CpG site?”</em>. We use <strong>regions</strong>
because it is well established that nearby CpG sites are <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6093082/" class="external-link">very likely
to share a similar DNAme profile</a> and therefore work as functional
units. Then, from a statistical point of view, testing separately
proximal CpGs that are part of the same unit is redundant. On the other
side, we use only <strong>variable</strong> regions because we are
interested in the units that display a high level of variability; in
other words, in non-variant sites there is no variability left to be
explained by genetics or environment. So, in conclusion, we use
<strong>VMRs</strong> to increase our power and reduce the multiple
hypothesis testing burden by grouping probes that are likely to work as
a biological unit, and by only focusing in the set of regions that are
of interest of this study.</p>
<p>RAMEN identifies 2 categories of VMRs:</p>
<ul>
<li>Canonical VMRs: Group of Highly Variable Probes that are proximal
and correlated. Highly Variable Probes are defined as probes above a
specific variance percentile threshold specified by the user (the
default is 90th percentile). The proximity distance and pearson
correlation threshold is specified by the user, and the defaults are 1
kilobase and 0.15 respectively. For guidance on which correlation
threshold to use, we recommend checking the Supplementary Figure 1 of
the CoMeBack R package (Gatev et al., 2020) where a simulation to
empirically determine a default guidance specification for a correlation
threshold parameter dependent on sample size is done.</li>
<li>Non canonical VMRs: Regions that are composed of Highly Variable
Probes that have no nearby probes measured in the array (according to
the distance parameter specified by the user). This category was created
to take into account the characteristics of the DNAme microarray
plataform, which has probes covering non-homogenelously the genome. This
is specially important for microarrays such as the EPIC array which has
a high number of probes in regulatory regions that are represented by a
single probe. Furthermore, these probes have been shown to be good
representatives of the methylation state of its surroundings (Pidsley et
al., 2016). By creating this category, we recover those informative HVPs
that would otherwise be excluded from the analysis because of working
with the canonical VMR definition in the array context.</li>
</ul>
<p>The first step is to identify <strong>Variable Methylated
Regions</strong>(VMRs) using the <code><a href="../reference/findVMRs.html">RAMEN::findVMRs()</a></code>
function. This function uses GenomicRanges::reduce() to group the
regions, which is strand-sensitive. In the Illumina microarrays, the
MAPINFO for all the probes is usually provided as for the + strand. If
you are using this array, we recommend to first convert the strand of
all the probes to “+”.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#We need to modify the RAMEN::test_array_manifest object by assigning to </span></span>
<span><span class="co">#row names to the probe ID column; it was saved this way because storing </span></span>
<span><span class="co">#the TargetID as row names reduced significantly the size of the data set. </span></span>
<span><span class="va">test_array_manifest_final</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_array_manifest.html">test_array_manifest</a></span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"TargetID"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">VMRs</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/findVMRs.html">findVMRs</a></span><span class="op">(</span>array_manifest <span class="op">=</span> <span class="va">test_array_manifest_final</span>, </span>
<span>                       methylation_data <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_methylation_data.html">test_methylation_data</a></span>, </span>
<span>                       cor_threshold <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                       var_method <span class="op">=</span> <span class="st">"variance"</span>, </span>
<span>                       var_threshold_percentile <span class="op">=</span> <span class="fl">0.9</span>, </span>
<span>                       max_distance <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Identifying Highly Variable Probes...</span></span>
<span><span class="co">#&gt; Identifying non canonical Variable Methylated Regions...</span></span>
<span><span class="co">#&gt; Identifying canonical Variable Methylated Regions...</span></span>
<span><span class="co">#&gt; Applying correlation filter to canonical Variable Methylated Regions...</span></span>
<span><span class="co">#&gt; Warning: executing %dopar% sequentially: no parallel backend registered</span></span>
<span></span>
<span><span class="co">#Take a look at the resulting object </span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">VMRs</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 4</span></span>
<span><span class="co">#&gt;  $ var_score_threshold   : Named num 13.5</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr "90%"</span></span>
<span><span class="co">#&gt;  $ highly_variable_probes:'data.frame':  300 obs. of  2 variables:</span></span>
<span><span class="co">#&gt;   ..$ TargetID : chr [1:300] "cg06187584" "cg09872009" "cg05437132" "cg26689799" ...</span></span>
<span><span class="co">#&gt;   ..$ var_score: num [1:300] 17.1 14.8 15.2 13.7 17.1 ...</span></span>
<span><span class="co">#&gt;  $ canonical_VMRs        :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots</span></span>
<span><span class="co">#&gt;   .. ..@ seqnames       :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ strand         :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ seqinfo        :Formal class 'Seqinfo' [package "GenomeInfoDb"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ elementMetadata:Formal class 'DFrame' [package "S4Vectors"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ elementType    : chr "ANY"</span></span>
<span><span class="co">#&gt;   .. ..@ metadata       : list()</span></span>
<span><span class="co">#&gt;  $ non_canonical_VMRs    :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots</span></span>
<span><span class="co">#&gt;   .. ..@ seqnames       :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ strand         :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ seqinfo        :Formal class 'Seqinfo' [package "GenomeInfoDb"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ elementMetadata:Formal class 'DFrame' [package "S4Vectors"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ elementType    : chr "ANY"</span></span>
<span><span class="co">#&gt;   .. ..@ metadata       : list()</span></span></code></pre></div>
<p>As we can see, <code><a href="../reference/findVMRs.html">RAMEN::findVMRs()</a></code> returns a list with
four elements:</p>
<ul>
<li>var_score_threshold: The MAD-score or variance threshold used to
define Highly Variable Probes.</li>
<li>highly_variable_probes: a data frame with the probes that passed the
variability score threshold imposed by the user, and their variability
score (MAD score or variance).</li>
<li>canonical_VMRs: a GRanges object with strict candidate VMRs -
regions composed of two or more contiguous, correlated and proximal
Highly Variable Probes; thresholds depend on the ones specified by the
user)</li>
<li>non_canonical_VMRs: a GRanges object with highly variable probes
without neighboring CpGs measured in max_distance on the array. Category
created to take into acccount the Illumina array design of single probes
capturing the methylation state of regulatory regions.</li>
</ul>
<p>Furthermore, we can see the following warning message in the chunk
above:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#&gt; Warning: executing %dopar% sequentially: no parallel backend registered</span></span></code></pre></div>
<p>This is printed in the screen just to warn us that
<code><a href="../reference/findVMRs.html">RAMEN::findVMRs()</a></code> is running sequentially. RAMEN supports
parallel computing for increased speed, which is really important when
working with real data sets that tend to contain information from the
whole genome. To do so, you have to set the parallel backend in your R
session BEFORE running the function (e.g.,
<code>doFuture::registerDoFuture()</code>) and then the evaluation
strategy (e.g., <code>future::plan(multisession)</code>). After that,
the function can be run normally. When working with big datasets, the
parallel backend might throw an error if you exceed the maximum allowed
size of globals exported for future expression. This can be fixed by
increasing the allowed size (e.g. running
<code>options(future.globals.maxSize= +Inf)</code>)</p>
<p>Finally, we will convert the output of <code><a href="../reference/findVMRs.html">RAMEN::findVMRs()</a></code>
to a data frame, which is an object that we can easily use to produce
plots and explore the results, and the object that is needed for the
following parts of the pipeline.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VMRs_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">VMRs</span><span class="op">[[</span><span class="st">"canonical_VMRs"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">VMRs</span><span class="op">[[</span><span class="st">"non_canonical_VMRs"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">width.1</span>,<span class="va">strand</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">VMRs_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;   seqnames    start      end width n_VMPs       probes median_correlation</span></span>
<span><span class="co">#&gt; 1       21 10990119 10990903   785      2 cg098720....          0.6099180</span></span>
<span><span class="co">#&gt; 2       21 11109021 11109336   316      2 cg007508....          0.6261681</span></span>
<span><span class="co">#&gt; 3       21 15456605 15456789   185      2 cg195094....          0.8595934</span></span>
<span><span class="co">#&gt; 4       21 15588038 15588830   793      2 cg023649....          0.6829510</span></span>
<span><span class="co">#&gt; 5       21 15955548 15956048   501      3 cg147721....          0.7548517</span></span>
<span><span class="co">#&gt; 6       21 16436438 16437768  1331      3 cg116304....          0.7124538</span></span></code></pre></div>
<p>With the VMRs as a data frame, we can explore our data set using
ggplot, such as the following example:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VMRs_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">width</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co">#Only plot canonical VMRs </span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">width</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">50</span>, fill <span class="op">=</span> <span class="st">"#BAB4D8"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Canonical VMRs width (bp)"</span><span class="op">)</span> </span></code></pre></div>
<div class="figure">
<img src="RAMEN_files/figure-html/unnamed-chunk-6-1.png" alt="Width of non canonical VMRs (base pairs)." width="700"><p class="caption">
Width of non canonical VMRs (base pairs).
</p>
</div>
<p>Next, we want to summarize the DNAme level of each VMR per
individual. To do this, we use <code><a href="../reference/summarizeVMRs.html">RAMEN::summarizeVMRs()</a></code>. For
non canonical VMRs, there is nothing to summarize, so the DNAme level of
the corresponding probe is returned. For canonical VMRs, the median
DNAme level of the region is returned per individual as the
representative value.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summarized_methyl_VMR</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/summarizeVMRs.html">summarizeVMRs</a></span><span class="op">(</span>VMRs_df <span class="op">=</span> <span class="va">VMRs_df</span>,</span>
<span>                                             methylation_data <span class="op">=</span> <span class="va">test_methylation_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Look at the resulting object</span></span>
<span><span class="va">summarized_methyl_VMR</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;            1        2        3        4           5</span></span>
<span><span class="co">#&gt; ID1 4.935942 2.853168 3.219038 5.384813 3.320883317</span></span>
<span><span class="co">#&gt; ID2 1.879166 2.699689 3.952487 2.927166 1.934283315</span></span>
<span><span class="co">#&gt; ID3 3.311818 1.078262 1.153573 4.690962 9.238518404</span></span>
<span><span class="co">#&gt; ID4 6.558106 4.683173 1.688233 3.448903 0.007576317</span></span>
<span><span class="co">#&gt; ID5 2.899969 4.930614 2.243714 5.788151 2.843904388</span></span></code></pre></div>
<p>The result is a data frame of VMR IDs as columns and individual IDs
as rows.</p>
</div>
<div class="section level3">
<h3 id="identify-cis-snps">Identify <em>cis</em> SNPs<a class="anchor" aria-label="anchor" href="#identify-cis-snps"></a>
</h3>
<p>After identifying the VMRs, we recommend to use only SNPs in
<em>cis</em> of each VMR, since genetic variants that associate with
DNAme changes tend to be more abundant in the surroundings of the
corresponding DNAme site (McClay <em>et al.</em>, 2015). Also, the
effect sizes of mQTLs (genetic variants associated with DNAme changes)
are stronger in <em>cis</em> SNPs compared to <em>trans</em> SNPs. Then,
by restricting the analysis to <em>cis</em> SNPs, we greatly reduce the
number of variables while keeping most of the important ones.</p>
<p>There is not a clear consensus on how close a SNP has to be from a
DNAme site to be considered <em>cis</em> - the distance threshold tend
to go from few kb to 1 megabase. We recommend to use a 1 Mb window to
cast a wide net to catch potentially relevant SNPs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VMRs_df</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/findCisSNPs.html">findCisSNPs</a></span><span class="op">(</span>VMRs_df <span class="op">=</span> <span class="va">VMRs_df</span>, </span>
<span>                            genotype_information <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_genotype_information.html">test_genotype_information</a></span>, </span>
<span>                            distance <span class="op">=</span> <span class="fl">1e+06</span><span class="op">)</span></span>
<span><span class="co">#&gt; Important: please make sure that the positions of the VMR data frame and the ones in the genotype information are from the same genome build.</span></span>
<span></span>
<span><span class="co">#Take a look at the result</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">VMRs_df</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Rows: 131</span></span>
<span><span class="co">#&gt; Columns: 10</span></span>
<span><span class="co">#&gt; $ VMR_index          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", …</span></span>
<span><span class="co">#&gt; $ seqnames           <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21,…</span></span>
<span><span class="co">#&gt; $ start              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 10990119, 11109021, 15456605, 15588038, 15955548, 1…</span></span>
<span><span class="co">#&gt; $ end                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 10990903, 11109336, 15456789, 15588830, 15956048, 1…</span></span>
<span><span class="co">#&gt; $ width              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 785, 316, 185, 793, 501, 1331, 888, 977, 573, 61, 5…</span></span>
<span><span class="co">#&gt; $ n_VMPs             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 2, 2, 2, 2, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, …</span></span>
<span><span class="co">#&gt; $ probes             <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> &lt;"cg09872009", "cg05437132"&gt;, &lt;"cg00750806", "cg12…</span></span>
<span><span class="co">#&gt; $ median_correlation <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.6099180, 0.6261681, 0.8595934, 0.6829510, 0.75485…</span></span>
<span><span class="co">#&gt; $ surrounding_SNPs   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 1, 1, 610, 667, 726, 794, 757, 529, 709, 788, 829, …</span></span>
<span><span class="co">#&gt; $ SNP                <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> "21:10873592:G:A", "21:10873592:G:A", &lt;"21:1459574…</span></span></code></pre></div>
<p>We can see that the resulting VMRs_df object is almost exactly the
same, but with two new columns (<em>surrounding_SNPs</em> and
<em>SNP</em>) that contain information about how many SNPs were found in
<em>cis</em> and what are their IDs according to the genotype data that
we have.</p>
<p>It is important to highlight the following characteristics of the
VMRs_df object:</p>
<ul>
<li>The <em>VMR_index</em> column is a character vector. This column
corresponds to the unique identifier of each VMR in our data set. It is
important to <strong>keep it as a character</strong> and be careful with
it not being converted to numeric, which can happen if you save the
VMRs_df object as a table, read it, and use that second object in the
rest of the pipeline.</li>
<li>The columns <em>probes</em> and <em>SNP</em> contain
<strong>lists</strong>. This structure is really important for the rest
of the analysis and columns containing lists will keep appearing in
other function outputs. If you want to know the recommended way to save
and load these objects, please check the [ Frequently Asked
Questions][].</li>
</ul>
<p>We can also explore the resulting object through plots such as the
following:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VMRs_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>type <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">n_VMPs</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"non canonical"</span>, <span class="co">#non canonical VMRs have only 1 probe by definition</span></span>
<span>                          <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"canonical"</span><span class="op">)</span>,</span>
<span>         surrounding_SNPs <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span> <span class="va">surrounding_SNPs</span> <span class="op">&gt;</span> <span class="fl">3000</span> <span class="op">~</span> <span class="fl">3000</span>,</span>
<span>                                       <span class="cn">TRUE</span> <span class="op">~</span> <span class="va">surrounding_SNPs</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">surrounding_SNPs</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="st">"type"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="RAMEN_files/figure-html/cissnps-1.png" alt="Disribution of SNPs in cis of VMRs." width="700"><p class="caption">
Disribution of SNPs in cis of VMRs.
</p>
</div>
</div>
<div class="section level3">
<h3 id="conduct-variable-selection-on-genome-and-exposome-variables">Conduct variable selection on genome and exposome variables<a class="anchor" aria-label="anchor" href="#conduct-variable-selection-on-genome-and-exposome-variables"></a>
</h3>
<p>The following stage in the pipeline is to screen the available
variables in our environmental and <em>cis</em> SNPs data sets to
identify the potentially relevant ones. This is achieved with the
<code><a href="../reference/selectVariables.html">RAMEN::selectVariables()</a></code> function. This function uses a
data-driven approach based on LASSO, which is an embedded variable
selection method commonly used in machine learning.</p>
<p>In a nutshell, LASSO penalizes models that are more complex (i.e.,
that contain more variables) in favor of simpler models (i.e. that
contain less variables), but not at the expense of reducing predictive
power. Using LASSO’s variable screening property (i.e., with high
probability, the LASSO estimated model includes the substantial
covariates and drops the redundant ones) this function is intended to
select genotype and environmental variables in each Variable Methylated
Region (VMR) with potential relevance in the presence of the
user-specified concomitant variables (which are known DNAme confounders
such as age, cell type proportion, etc.). For more information about the
method, we encourage the users to read the documentation of the
function, and for further information about LASSO we direct readers to
Bühlmann and Van de Geer, 2011.</p>
<p>Overall, conducting our variable selection strategy reduces the
overall computational time and improves the modeling performance by:</p>
<ul>
<li>Reducing the universe of variables that will be used to fit models
in the following stage (G/E/G+E/GxE model fitting and comparison)</li>
<li>Removing redundant variables, which are highly expected in genetic
and environmental data sets with a high number of variables</li>
<li>Limiting the interactions terms to scenarios where both the G and E
main effects were selected to be potentially relevant, which can be
think of as an interaction variable selection using a weak
hierarchy</li>
<li>Using LASSO, a method with good variable selection performance and
scalability</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_variables</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/selectVariables.html">selectVariables</a></span><span class="op">(</span></span>
<span>  VMRs_df <span class="op">=</span> <span class="va">VMRs_df</span>,</span>
<span>  genotype_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_genotype_matrix.html">test_genotype_matrix</a></span>,</span>
<span>  environmental_matrix<span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_environmental_matrix.html">test_environmental_matrix</a></span>,</span>
<span>  covariates <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_covariates.html">test_covariates</a></span>,</span>
<span>  summarized_methyl_VMR <span class="op">=</span> <span class="va">summarized_methyl_VMR</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: foreach</span></span>
<span><span class="co">#&gt; Loading required package: rngtools</span></span></code></pre></div>
<p>Since LASSO makes use of Random Number Generation, setting a seed is
highly encouraged for result’s reproducibility using the <em>seed</em>
argument. As a note, setting a seed inside of this function modifies the
seed globally (which is R’s default behavior).</p>
<p>The output of <code><a href="../reference/selectVariables.html">RAMEN::selectVariables()</a></code> is an object with
the VMR index, and the G and E variables selected for each VMR.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">selected_variables</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 131</span></span>
<span><span class="co">#&gt; Columns: 3</span></span>
<span><span class="co">#&gt; $ VMR_index      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"…</span></span>
<span><span class="co">#&gt; $ selected_genot <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> "21:10873592:G:A", "21:10873592:G:A", &lt;"21:15450190:C:…</span></span>
<span><span class="co">#&gt; $ selected_env   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> &lt;"E43", "E3", "E5", "E7", "E24", "E25", "E28", "E35", …</span></span></code></pre></div>
<p>We can see how using <code><a href="../reference/selectVariables.html">RAMEN::selectVariables()</a></code> reduces
the number of variables (originally 100 environmental variables and 785
SNPs per VMR on average as seen in Figure @ref(fig:cissnps)).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_variables</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">VMRs_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>              <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">VMR_index</span>,<span class="va">n_VMPs</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            by <span class="op">=</span> <span class="st">"VMR_index"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/transmute.html" class="external-link">transmute</a></span><span class="op">(</span>VMR_index <span class="op">=</span> <span class="va">VMR_index</span>,</span>
<span>            VMR_type <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">n_VMPs</span> <span class="op">&gt;</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"canonical"</span>,</span>
<span>                          <span class="va">n_VMPs</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"non canonical"</span><span class="op">)</span>, </span>
<span>            genotype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">selected_genot</span><span class="op">)</span>,</span>
<span>            environment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">selected_env</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">VMR_index</span>, <span class="va">VMR_type</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">name</span>,</span>
<span>         variables <span class="op">=</span> <span class="va">value</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">VMR_type</span>, y <span class="op">=</span> <span class="va">variables</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html" class="external-link">geom_violin</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>width<span class="op">=</span><span class="fl">0.1</span>, outlier.shape<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">group</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Selected variables"</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="RAMEN_files/figure-html/selectedvars-1.png" alt="Number of G and E selected variables." width="700"><p class="caption">
Number of G and E selected variables.
</p>
</div>
<p>It is also expected in real data to have VMRs where no SNP and/or no
environmental variables were selected, since not all the DNAme sites in
the genome are expected to show an association with the genetic
variation or environmental exposures data sets that are captured in a
study. The proportion of VMRs under these scenarios will depend on the
data sets.</p>
<div class="section level4">
<h4 id="authors-note-about-interpretation">Author’s note about interpretation<a class="anchor" aria-label="anchor" href="#authors-note-about-interpretation"></a>
</h4>
<p>LASSO variable selection is not consistent when there is
multicollinearity in the data (i.e., correlation between variables),
which is expected due to the high amount of G and E variables that are
present in studies of this kind. This means that if you were to run
LASSO several times, and two variables were to be highly correlated, the
method would select one and drop the other one at random. This is not a
problem with the pipeline because the main conclusion per VMR is whether
the DNAme is better explained by G and/or E components. As an example,
if a VMR is better explained by SNP1 and SNP2, which are both highly
correlated one with the other, LASSO will randomly pick SNP1 OR SNP2
(because they are relevant but they provide redundant information); if
we were to fit a model with SNP1 or SNP2 in the following stage, the
winning model would still be G. In other words, the main goal of the
pipeline is to know whether the VMR’s DNAme is better explained by G
and/or E. The user is therefore warned to <strong>be cautious not to
over-interpret the individual selected variables</strong>. Selected
variables might be used as hypothesis generators of associations,
keeping in mind that the selected variable might be representing other
variables in the data set that provide similar information.</p>
</div>
</div>
<div class="section level3">
<h3 id="identify-the-best-explanatory-model-gegegxe-per-vmr">Identify the best explanatory model (G/E/G+E/GxE) per VMR<a class="anchor" aria-label="anchor" href="#identify-the-best-explanatory-model-gegegxe-per-vmr"></a>
</h3>
<p>Now that we have selected the list of potentially relevant G and E
variables, we will fit the models mentioned in Table
@ref(tab:modelstable) using the <code><a href="../reference/lmGE.html">RAMEN::lmGE()</a></code> function.
This function fits for each VMR G and E models with all of the variables
selected, as well as all their possible pairwise combinations of G+E and
GxE.</p>
<p>After fitting this model, the best model per group (group = G, E, G+E
or GxE) is selected using Akaike Information Criterion (AIC) or Bayesian
Information Criterion (BIC). We recommend using AIC because BIC assumes
that the true model is in the set of compared models. Since this
function fits models with individual variables, and we assume that DNAme
variability is more likely to be influenced by more than one single
SNP/environmental exposure at a time, we hypothesize that in most cases,
the true model will not be in the set of compared models. Also, AIC
excels in situations where all models in the model space are
“incorrect”, and AIC is preferentially used in cases where the true
underlying function is unknown and our selected model could belong to a
very large class of functions where the relationship could be pretty
complex. It is worth mentioning however that, both metrics tend to pick
the same model in a large number of scenarios. We suggest the users to
read Arijit Chakrabarti &amp; Jayanta K. Ghosh, 2011 for further
information about the difference between these metrics. After selecting
the best model per group (G,E,G+E pr GxE), the model with the lowest AIC
will be declared as the winning model.</p>
<p>Additionally, <code><a href="../reference/lmGE.html">RAMEN::lmGE()</a></code> conducts a variance
decomposition analysis, so that the relative R2 contribution of each of
the variables of interest (G, E and GxE) is reported. This decomposition
is done using the <em><a href="https://CRAN.R-project.org/package=relaimpo" class="external-link">relaimpo</a></em> R
package, using the Lindeman, Merenda and Gold (lmg) method, which is
based on the heuristic approach of averaging the relative R contribution
of each variable over all input orders in the linear model.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmge_res</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/lmGE.html">lmGE</a></span><span class="op">(</span></span>
<span>  selected_variables <span class="op">=</span> <span class="va">selected_variables</span>,</span>
<span>  summarized_methyl_VMR <span class="op">=</span> <span class="va">summarized_methyl_VMR</span>,</span>
<span>  genotype_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_genotype_matrix.html">test_genotype_matrix</a></span>,</span>
<span>  environmental_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_environmental_matrix.html">test_environmental_matrix</a></span>,</span>
<span>  covariates <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_covariates.html">test_covariates</a></span>,</span>
<span>  model_selection <span class="op">=</span> <span class="st">"AIC"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check the output</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">lmge_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 128</span></span>
<span><span class="co">#&gt; Columns: 13</span></span>
<span><span class="co">#&gt; $ VMR_index       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11…</span></span>
<span><span class="co">#&gt; $ model_group     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "G+E", "G+E", "GxE", "GxE", "G+E", "GxE", "G+E", "GxE"…</span></span>
<span><span class="co">#&gt; $ variables       <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> &lt;"21:10873592:G:A", "E43"&gt;, &lt;"21:10873592:G:A", "E43"…</span></span>
<span><span class="co">#&gt; $ tot_r_squared   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.5527507, 0.5115764, 0.6075657, 0.6794845, 0.7680461,…</span></span>
<span><span class="co">#&gt; $ g_r_squared     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.1996156, 0.2095645, 0.3064024, 0.2738786, 0.3981840,…</span></span>
<span><span class="co">#&gt; $ e_r_squared     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.34206198, 0.27143559, 0.15293109, 0.20266291, 0.2769…</span></span>
<span><span class="co">#&gt; $ gxe_r_squared   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> NA, NA, 0.1330680, 0.1126333, NA, 0.1432113, NA, 0.108…</span></span>
<span><span class="co">#&gt; $ AIC             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 144.6841, 148.7250, 145.6987, 135.7819, 129.9252, 123.…</span></span>
<span><span class="co">#&gt; $ second_winner   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "GxE", "GxE", "G+E", "G+E", "GxE", "G+E", "GxE", "G+E"…</span></span>
<span><span class="co">#&gt; $ delta_aic       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 1.04970268, 1.34557464, 1.01022314, 3.88212651, 0.0761…</span></span>
<span><span class="co">#&gt; $ delta_r_squared <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> -0.013945293, -0.010539189, 0.041420496, 0.069427874, …</span></span>
<span><span class="co">#&gt; $ basal_AIC       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 164.4893, 165.2905, 167.3019, 161.0771, 166.8374, 159.…</span></span>
<span><span class="co">#&gt; $ basal_rsquared  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.011073059, 0.030576374, 0.015164242, 0.090309667, 0.…</span></span></code></pre></div>
<div class="section level4">
<h4 id="remove-poor-performing-winning-models">Remove poor performing winning models<a class="anchor" aria-label="anchor" href="#remove-poor-performing-winning-models"></a>
</h4>
<p>After identifying the winning models per VMR, we proceed to the last
step of the pipeline which is computing a null distribution to remove
the winning models that are very likely to be winners by chance.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">null_dist</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/nullDistGE.html">nullDistGE</a></span><span class="op">(</span></span>
<span>  VMRs_df <span class="op">=</span> <span class="va">VMRs_df</span>,</span>
<span>  genotype_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_genotype_matrix.html">test_genotype_matrix</a></span>,</span>
<span>  environmental_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_environmental_matrix.html">test_environmental_matrix</a></span>,</span>
<span>  summarized_methyl_VMR <span class="op">=</span> <span class="va">summarized_methyl_VMR</span>,</span>
<span>  permutations <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  covariates <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_covariates.html">test_covariates</a></span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  model_selection <span class="op">=</span> <span class="st">"AIC"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Take a look at the object </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">null_dist</span><span class="op">)</span></span>
<span><span class="co">#&gt;   VMR_index tot_r_squared model_group R2_difference AIC_difference permutation</span></span>
<span><span class="co">#&gt; 1         1     0.2606017           E     0.2495286       157.7547           1</span></span>
<span><span class="co">#&gt; 2         2     0.2976822           E     0.2671059       157.5905           1</span></span>
<span><span class="co">#&gt; 3         3     0.6086939         G+E     0.5935297       143.5972           1</span></span>
<span><span class="co">#&gt; 4         4     0.3777878           E     0.2874781       151.5921           1</span></span>
<span><span class="co">#&gt; 5         6     0.6133220         G+E     0.5269145       137.8163           1</span></span>
<span><span class="co">#&gt; 6         7     0.3451632           E     0.2497069       155.1677           1</span></span></code></pre></div>
<p>Now we will get a cutoff for single (G,E) and joint (G+E,GxE)
models.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a cutoff of the 95th percentile of the null distribution</span></span>
<span><span class="va">null_dist</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">R2_difference</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="st">"model_group"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre></div>
<div class="figure">
<img src="RAMEN_files/figure-html/unnamed-chunk-13-1.png" alt="R2 difference (winner - basal) in a suffled data set." width="700"><p class="caption">
R2 difference (winner - basal) in a suffled data set.
</p>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Get a cutoff of the 95th percentile of the null distribution for single and joint models</span></span>
<span><span class="va">cutoff_single</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">null_dist</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>                           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">model_group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G"</span>,<span class="st">"E"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>                           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">R2_difference</span><span class="op">)</span>, </span>
<span>                         <span class="fl">0.95</span><span class="op">)</span></span>
<span><span class="va">cutoff_joint</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">null_dist</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>                           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">model_group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G+E"</span>,<span class="st">"GxE"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>                           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">R2_difference</span><span class="op">)</span>, </span>
<span>                         <span class="fl">0.95</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">final_res</span> <span class="op">=</span> <span class="va">lmge_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>r2_difference_basal <span class="op">=</span> <span class="va">tot_r_squared</span> <span class="op">-</span> <span class="va">basal_rsquared</span>,</span>
<span>         pass_cutoff_threshold <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">model_group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G"</span>, <span class="st">"E"</span><span class="op">)</span> <span class="op">~</span> <span class="va">r2_difference_basal</span> <span class="op">&gt;</span> <span class="va">cutoff_single</span>,</span>
<span>                                           <span class="va">model_group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G+E"</span>, <span class="st">"GxE"</span><span class="op">)</span> <span class="op">~</span> <span class="va">r2_difference_basal</span> <span class="op">&gt;</span> <span class="va">cutoff_joint</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pass_cutoff_threshold</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co">#Filter based on the cutoff threshold</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">pass_cutoff_threshold</span><span class="op">)</span> <span class="co">#Drop transient column</span></span>
<span></span>
<span><span class="co">#Check the final results</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">final_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 7</span></span>
<span><span class="co">#&gt; Columns: 14</span></span>
<span><span class="co">#&gt; $ VMR_index           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "5", "6", "28", "39", "96", "98", "106"</span></span>
<span><span class="co">#&gt; $ model_group         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "G+E", "GxE", "E", "GxE", "G", "G", "GxE"</span></span>
<span><span class="co">#&gt; $ variables           <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> &lt;"21:15671097:T:A", "E43"&gt;, &lt;"21:15671097:T:A", "E…</span></span>
<span><span class="co">#&gt; $ tot_r_squared       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.7680461, 0.7769276, 0.5555035, 0.7097936, 0.6431…</span></span>
<span><span class="co">#&gt; $ g_r_squared         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.39818402, 0.38220942, NA, 0.09101486, 0.6398122…</span></span>
<span><span class="co">#&gt; $ e_r_squared         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.2769884, 0.1650993, 0.5244324, 0.2406247, NA, NA…</span></span>
<span><span class="co">#&gt; $ gxe_r_squared       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> NA, 0.1432113, NA, 0.3522698, NA, NA, 0.1515714</span></span>
<span><span class="co">#&gt; $ AIC                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 129.9252, 123.3999, 145.1528, 147.3162, 140.8549, …</span></span>
<span><span class="co">#&gt; $ second_winner       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "GxE", "G+E", NA, "G+E", "G+E", NA, "G+E"</span></span>
<span><span class="co">#&gt; $ delta_aic           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.07612103, 3.90964213, NA, 3.27939276, 0.34990027…</span></span>
<span><span class="co">#&gt; $ delta_r_squared     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> -0.01440811, 0.04856945, NA, 0.05583976, -0.019097…</span></span>
<span><span class="co">#&gt; $ basal_AIC           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 166.8374, 159.6965, 166.5303, 177.6443, 169.6680, …</span></span>
<span><span class="co">#&gt; $ basal_rsquared      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.092873702, 0.086407501, 0.031071178, 0.025884145…</span></span>
<span><span class="co">#&gt; $ r2_difference_basal <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.6751724, 0.6905201, 0.5244324, 0.6839094, 0.6398…</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] doRNG_1.8.6      rngtools_1.5.2   foreach_1.5.2    tidyr_1.3.0     </span></span>
<span><span class="co">#&gt; [5] ggplot2_3.4.4    dplyr_1.1.3      RAMEN_0.1.9000   knitr_1.45      </span></span>
<span><span class="co">#&gt; [9] BiocStyle_2.28.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] shape_1.4.6             gtable_0.3.4            xfun_0.41              </span></span>
<span><span class="co">#&gt;  [4] bslib_0.5.1             lattice_0.21-9          vctrs_0.6.4            </span></span>
<span><span class="co">#&gt;  [7] tools_4.3.2             bitops_1.0-7            generics_0.1.3         </span></span>
<span><span class="co">#&gt; [10] stats4_4.3.2            parallel_4.3.2          tibble_3.2.1           </span></span>
<span><span class="co">#&gt; [13] fansi_1.0.5             highr_0.10              pkgconfig_2.0.3        </span></span>
<span><span class="co">#&gt; [16] Matrix_1.6-1.1          desc_1.4.2              S4Vectors_0.38.2       </span></span>
<span><span class="co">#&gt; [19] lifecycle_1.0.4         GenomeInfoDbData_1.2.10 compiler_4.3.2         </span></span>
<span><span class="co">#&gt; [22] farver_2.1.1            stringr_1.5.1           textshaping_0.3.7      </span></span>
<span><span class="co">#&gt; [25] munsell_0.5.0           mitools_2.4             codetools_0.2-19       </span></span>
<span><span class="co">#&gt; [28] survey_4.2-1            GenomeInfoDb_1.36.4     htmltools_0.5.7        </span></span>
<span><span class="co">#&gt; [31] sass_0.4.7              RCurl_1.98-1.13         yaml_2.3.7             </span></span>
<span><span class="co">#&gt; [34] glmnet_4.1-8            pillar_1.9.0            pkgdown_2.0.7          </span></span>
<span><span class="co">#&gt; [37] jquerylib_0.1.4         MASS_7.3-60             cachem_1.0.8           </span></span>
<span><span class="co">#&gt; [40] iterators_1.0.14        boot_1.3-28.1           tidyselect_1.2.0       </span></span>
<span><span class="co">#&gt; [43] digest_0.6.33           stringi_1.8.1           purrr_1.0.2            </span></span>
<span><span class="co">#&gt; [46] bookdown_0.36           splines_4.3.2           labeling_0.4.3         </span></span>
<span><span class="co">#&gt; [49] rprojroot_2.0.4         fastmap_1.1.1           grid_4.3.2             </span></span>
<span><span class="co">#&gt; [52] colorspace_2.1-0        cli_3.6.1               magrittr_2.0.3         </span></span>
<span><span class="co">#&gt; [55] survival_3.5-7          utf8_1.2.4              corpcor_1.6.10         </span></span>
<span><span class="co">#&gt; [58] withr_2.5.2             scales_1.2.1            rmarkdown_2.25         </span></span>
<span><span class="co">#&gt; [61] XVector_0.40.0          ragg_1.2.6              memoise_2.0.1          </span></span>
<span><span class="co">#&gt; [64] evaluate_0.23           GenomicRanges_1.52.1    IRanges_2.34.1         </span></span>
<span><span class="co">#&gt; [67] rlang_1.1.2             Rcpp_1.0.11             DBI_1.1.3              </span></span>
<span><span class="co">#&gt; [70] glue_1.6.2              BiocManager_1.30.22     BiocGenerics_0.46.0    </span></span>
<span><span class="co">#&gt; [73] jsonlite_1.8.7          R6_2.5.1                relaimpo_2.2-7         </span></span>
<span><span class="co">#&gt; [76] systemfonts_1.0.5       fs_1.6.3                zlibbioc_1.46.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Erick I. Navarro-Delgado.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
