<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>RAMEN • RAMEN</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="RAMEN">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">RAMEN</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/RAMEN.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>RAMEN</h1>
                        <h4 data-toc-skip class="author">Erick I.
Navarro-Delgado</h4>
            <address class="author_afil">
      The University of British
Columbia<br><a class="author_email" href="mailto:#"></a><a href="mailto:erick.navarrodelgado@bcchr.ca" class="email">erick.navarrodelgado@bcchr.ca</a>
      </address>
                  
      

      <div class="d-none name"><code>RAMEN.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>Regional Association of Methylome variability with the
Exposome and geNome (RAMEN)</strong> is an R package whose goal is to
integrate genomic, methylomic and exposomic data to model the
contribution of genetics (G) and the environment (E) to DNA methylation
(DNAme) variability. RAMEN identifies Variable Methylated Regions (VMRs)
in microarray DNAme data and then, using genotype and environmental
data, it identifies which of the following models better explains this
variability in regions across the methylome:</p>
<table class="table">
<caption>Fitted models</caption>
<thead><tr class="header">
<th align="left">Model</th>
<th align="left">Name</th>
<th align="left">Abbreviation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">DNAme ~ G + covars</td>
<td align="left">Genetics</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="left">DNAme ~ E + covars</td>
<td align="left">Environmental exposure</td>
<td align="left">E</td>
</tr>
<tr class="odd">
<td align="left">DNAme ~ G + E + covars</td>
<td align="left">Additive</td>
<td align="left">G+E</td>
</tr>
<tr class="even">
<td align="left">DNAme ~ G + E + G*E + covars</td>
<td align="left">Interaction</td>
<td align="left">GxE</td>
</tr>
</tbody>
</table>
<p>where G variables are represented by SNPs, E variables by
environmental exposures, and where covars are concomitant variables
(i.e. variables that are adjusted for in the model and not of interest
in the study such as cell type proportion, age, etc.).</p>
<p>The main <a href="#gene-environment-interaction-analysis">gene-environment
interaction modeling</a> pipeline is conducted though six core
functions:</p>
<ul>
<li>
<code><a href="../reference/findVMRs.html">findVMRs()</a></code> identifies Variable Methylated Regions
(VMRs) in microarrays</li>
<li>
<code><a href="../reference/summarizeVMRs.html">summarizeVMRs()</a></code>summarizes the regional methylation
state of each VMR</li>
<li>
<code><a href="../reference/findCisSNPs.html">findCisSNPs()</a></code> identifies the SNPs in <em>cis</em> of
each VMR</li>
<li>
<code><a href="../reference/selectVariables.html">selectVariables()</a></code> conducts a LASSO-based variable
selection strategy to identify potentially relevant <em>cis</em> SNPs
and environmental variables</li>
<li>
<code><a href="../reference/lmGE.html">lmGE()</a></code> fits linear single-variable genetic (G) and
environmental (E), and pairwise additive (G+E) and interaction (GxE)
linear models and select the best explanatory model per VMR.</li>
<li>
<code><a href="../reference/nullDistGE.html">nullDistGE()</a></code> simulates a delta R squared null
distribution of G and E effects on DNAme variability. Useful for
filtering out poor-performing best explanatory models selected by
<em>lmGE()</em>.</li>
</ul>
<p>These functions are compatible with parallel computing, which is
recommended due to the computationally intensive tasks conducted by the
package.</p>
<p>In addition to the <a href="#gene-environment-interaction-analysis">standard gene-environment
interaction modeling pipeline</a>, RAMEN can be useful for other DNAme
analyses (see <a href="#variations-to-the-standard-workflow">Variations
to the standard workflow</a>), such as reducing the tested sites in
Epigenome Wide Association Studies, grouping DNAme probes into regions,
identifying SNPs near a probe, etc.</p>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>The manuscript detailing RAMEN and its use is currently under
preparation. For more information about this please contact Erick I.
Navarro-Delgado at <a href="mailto:erick.navarrodelgado@bcchr.ca" class="email">erick.navarrodelgado@bcchr.ca</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="gene-environment-interaction-analysis">Gene-environment interaction analysis<a class="anchor" aria-label="anchor" href="#gene-environment-interaction-analysis"></a>
</h2>
<p>The main purpose of the RAMEN package is to conduct a methylome-wide
analysis to identify which model (G, E, G+E or GxE) better explains the
variability across the genome. In this vignette, we will illustrate how
to use the package.</p>
<p>To conduct this analysis, the following cleaned data sets (i.e. after
quality and exploratory data analysis checks) from a cohort are
required:</p>
<ul>
<li>DNAme data</li>
<li>DNAme array manifest</li>
<li>Genotyping data</li>
<li>Genotype information</li>
<li>Environmental exposure data</li>
<li>Concomitant variables data</li>
</ul>
<p>Once that we have that data, the overview of the pipeline is the
following:</p>
<div class="float">
<img src="RAMEN_pipeline.png" alt="RAMEN pipeline"><div class="figcaption">RAMEN pipeline</div>
</div>
<p>where:</p>
<ul>
<li>DNAme data is grouped into VMRs, and then the DNAme state per
individual is summarized in each region</li>
<li>Using the identified VMRs and the genomic information, we identify
the SNPs in <em>cis</em> for each VMR</li>
<li>Both the <em>cis</em> SNPs and the exposome data are subjected to
the variable selection stage</li>
<li>The selected variables (SNPs a and Es) enter the modelling stage,
which outputs one single winning model per VMR</li>
<li>The thresholds obtained from the simulated null distribution are
used to remove winning models which performance are likely to be due to
chance.</li>
</ul>
<p>In the following sections we will go through each of these steps and
guide the user regarding the recommended parameters to use in each
function of the package. For illustration purposes, we provide small toy
data sets that do not intend to simulate a real biological phenomenon.
These data sets are already available in the RAMEN package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Load the packages used throughout the vignette</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ericknavarrod.github.io/RAMEN/">RAMEN</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="identify-vmrs-and-summarize-their-methylation-state">Identify VMRs and summarize their methylation state<a class="anchor" aria-label="anchor" href="#identify-vmrs-and-summarize-their-methylation-state"></a>
</h3>
<p>The first step of the pipeline is to identify the <strong>Variable
Methylated Regions</strong>(VMRs) in the data set. You might be
wondering <em>“What is a VMR and why do we use them instead of DNAme
levels from each CpG site?”</em>. We use <strong>regions</strong>
because it is well established that nearby CpG sites are <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6093082/" class="external-link">very likely
to share a similar DNAme profile</a> and therefore work as functional
units. Then, from a statistical point of view, testing separately
proximal CpGs that are part of the same unit is redundant. On the other
side, we use only <strong>variable</strong> regions because we are
interested in the units that display a high level of variability; in
other words, in non-variant sites there is no variability left to be
explained by genetics or environment. So, in conclusion, we use
<strong>VMRs</strong> to increase our power and reduce the multiple
hypothesis testing burden by grouping probes that are likely to work as
a biological unit, and by only focusing in the set of regions that are
of interest of this study.</p>
<p>RAMEN identifies 2 categories of VMRs:</p>
<ul>
<li>Canonical VMRs: Group of Highly Variable Probes that are proximal
and correlated. Highly Variable Probes are defined as probes above a
specific variance percentile threshold specified by the user (the
default is 90th percentile). The proximity distance and pearson
correlation threshold is specified by the user, and the defaults are 1
kilobase and 0.15 respectively. For guidance on which correlation
threshold to use, we recommend checking the Supplementary Figure 1 of
the CoMeBack R package (Gatev et al., 2020) where a simulation to
empirically determine a default guidance specification for a correlation
threshold parameter dependent on sample size is done.</li>
<li>Non canonical VMRs: Regions that are composed of Highly Variable
Probes that have no nearby probes measured in the array (according to
the distance parameter specified by the user). This category was created
to take into account the characteristics of the DNAme microarray
plataform, which has probes covering non-homogenelously the genome. This
is specially important for microarrays such as the EPIC array which has
a high number of probes in regulatory regions that are represented by a
single probe. Furthermore, these probes have been shown to be good
representatives of the methylation state of its surroundings (Pidsley et
al., 2016). By creating this category, we recover those informative HVPs
that would otherwise be excluded from the analysis because of working
with the canonical VMR definition in the array context.</li>
</ul>
<p>The first step is to identify <strong>Variable Methylated
Regions</strong>(VMRs) using the <code><a href="../reference/findVMRs.html">RAMEN::findVMRs()</a></code>
function. This function uses GenomicRanges::reduce() to group the
regions, which is strand-sensitive. In the Illumina microarrays, the
MAPINFO for all the probes is usually provided as for the + strand. If
you are using this array, we recommend to first convert the strand of
all the probes to “+”.</p>
<p>For this step, we recommend users to use M-values because its use is
more appropiate for statistical analyses (see Pan Du, <em>et al.</em>,
2010, <em>BMC Bioinformatics</em>)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#We need to modify the RAMEN::test_array_manifest object by assigning to </span></span>
<span><span class="co">#row names to the probe ID column; it was saved this way because storing </span></span>
<span><span class="co">#the TargetID as row names reduced significantly the size of the data set. </span></span>
<span><span class="va">test_array_manifest_final</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_array_manifest.html">test_array_manifest</a></span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"TargetID"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">VMRs</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/findVMRs.html">findVMRs</a></span><span class="op">(</span>array_manifest <span class="op">=</span> <span class="va">test_array_manifest_final</span>, </span>
<span>                       methylation_data <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_methylation_data.html">test_methylation_data</a></span>, </span>
<span>                       cor_threshold <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                       var_method <span class="op">=</span> <span class="st">"variance"</span>, </span>
<span>                       var_threshold_percentile <span class="op">=</span> <span class="fl">0.9</span>, </span>
<span>                       max_distance <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Identifying Highly Variable Probes...</span></span>
<span><span class="co">#&gt; Identifying non canonical Variable Methylated Regions...</span></span>
<span><span class="co">#&gt; Identifying canonical Variable Methylated Regions...</span></span>
<span><span class="co">#&gt; Applying correlation filter to canonical Variable Methylated Regions...</span></span>
<span><span class="co">#&gt; Warning: executing %dopar% sequentially: no parallel backend registered</span></span>
<span></span>
<span><span class="co">#Take a look at the resulting object </span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">VMRs</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 4</span></span>
<span><span class="co">#&gt;  $ var_score_threshold   : Named num 13.5</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr "90%"</span></span>
<span><span class="co">#&gt;  $ highly_variable_probes:'data.frame':  300 obs. of  2 variables:</span></span>
<span><span class="co">#&gt;   ..$ TargetID : chr [1:300] "cg06187584" "cg09872009" "cg05437132" "cg26689799" ...</span></span>
<span><span class="co">#&gt;   ..$ var_score: num [1:300] 17.1 14.8 15.2 13.7 17.1 ...</span></span>
<span><span class="co">#&gt;  $ canonical_VMRs        :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots</span></span>
<span><span class="co">#&gt;   .. ..@ seqnames       :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ strand         :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ seqinfo        :Formal class 'Seqinfo' [package "GenomeInfoDb"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ elementMetadata:Formal class 'DFrame' [package "S4Vectors"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ elementType    : chr "ANY"</span></span>
<span><span class="co">#&gt;   .. ..@ metadata       : list()</span></span>
<span><span class="co">#&gt;  $ non_canonical_VMRs    :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots</span></span>
<span><span class="co">#&gt;   .. ..@ seqnames       :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ strand         :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ seqinfo        :Formal class 'Seqinfo' [package "GenomeInfoDb"] with 4 slots</span></span>
<span><span class="co">#&gt;   .. ..@ elementMetadata:Formal class 'DFrame' [package "S4Vectors"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ elementType    : chr "ANY"</span></span>
<span><span class="co">#&gt;   .. ..@ metadata       : list()</span></span></code></pre></div>
<p>As we can see, <code><a href="../reference/findVMRs.html">RAMEN::findVMRs()</a></code> returns a list with
four elements:</p>
<ul>
<li>var_score_threshold: The MAD-score or variance threshold used to
define Highly Variable Probes.</li>
<li>highly_variable_probes: a data frame with the probes that passed the
variability score threshold imposed by the user, and their variability
score (MAD score or variance).</li>
<li>canonical_VMRs: a GRanges object with strict candidate VMRs -
regions composed of two or more contiguous, correlated and proximal
Highly Variable Probes; thresholds depend on the ones specified by the
user)</li>
<li>non_canonical_VMRs: a GRanges object with highly variable probes
without neighboring CpGs measured in max_distance on the array. Category
created to take into acccount the Illumina array design of single probes
capturing the methylation state of regulatory regions.</li>
</ul>
<p>Furthermore, we can see the following warning message in the chunk
above:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#&gt; Warning: executing %dopar% sequentially: no parallel backend registered</span></span></code></pre></div>
<p>This is printed in the screen just to warn us that
<code><a href="../reference/findVMRs.html">RAMEN::findVMRs()</a></code> is running sequentially. RAMEN supports
parallel computing for increased speed, which is really important when
working with real data sets that tend to contain information from the
whole genome. To do so, you have to set the parallel backend in your R
session BEFORE running the function (e.g.,
<code>doFuture::registerDoFuture()</code>) and then the evaluation
strategy (e.g., <code>future::plan(multisession)</code>). After that,
the function can be run normally. When working with big datasets, the
parallel backend might throw an error if you exceed the maximum allowed
size of globals exported for future expression. This can be fixed by
increasing the allowed size (e.g. running
<code>options(future.globals.maxSize= +Inf)</code>)</p>
<p>Finally, we will convert the output of <code><a href="../reference/findVMRs.html">RAMEN::findVMRs()</a></code>
to a data frame, which is an object that we can easily use to produce
plots and explore the results, and the object that is needed for the
following parts of the pipeline.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VMRs_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">VMRs</span><span class="op">[[</span><span class="st">"canonical_VMRs"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">VMRs</span><span class="op">[[</span><span class="st">"non_canonical_VMRs"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">width.1</span>,<span class="va">strand</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">VMRs_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;   seqnames    start      end width n_VMPs       probes median_correlation</span></span>
<span><span class="co">#&gt; 1       21 10990119 10990903   785      2 cg098720....          0.6099180</span></span>
<span><span class="co">#&gt; 2       21 11109021 11109336   316      2 cg007508....          0.6261681</span></span>
<span><span class="co">#&gt; 3       21 15456605 15456789   185      2 cg195094....          0.8595934</span></span>
<span><span class="co">#&gt; 4       21 15588038 15588830   793      2 cg023649....          0.6829510</span></span>
<span><span class="co">#&gt; 5       21 15955548 15956048   501      3 cg147721....          0.7548517</span></span>
<span><span class="co">#&gt; 6       21 16436438 16437768  1331      3 cg116304....          0.7124538</span></span></code></pre></div>
<p>With the VMRs as a data frame, we can explore our data set using
ggplot, such as the following example:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VMRs_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">width</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co">#Only plot canonical VMRs </span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">width</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">50</span>, fill <span class="op">=</span> <span class="st">"#BAB4D8"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Canonical VMRs width (bp)"</span><span class="op">)</span> </span></code></pre></div>
<div class="figure">
<img src="RAMEN_files/figure-html/unnamed-chunk-6-1.png" alt="Width of non canonical VMRs (base pairs)." width="700"><p class="caption">
Width of non canonical VMRs (base pairs).
</p>
</div>
<p>Next, we want to summarize the DNAme level of each VMR per
individual. To do this, we use <code><a href="../reference/summarizeVMRs.html">RAMEN::summarizeVMRs()</a></code>. For
non canonical VMRs, there is nothing to summarize, so the DNAme level of
the corresponding probe is returned. For canonical VMRs, the median
DNAme level of the region is returned per individual as the
representative value.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summarized_methyl_VMR</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/summarizeVMRs.html">summarizeVMRs</a></span><span class="op">(</span>VMRs_df <span class="op">=</span> <span class="va">VMRs_df</span>,</span>
<span>                                             methylation_data <span class="op">=</span> <span class="va">test_methylation_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Look at the resulting object</span></span>
<span><span class="va">summarized_methyl_VMR</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;            1        2        3        4           5</span></span>
<span><span class="co">#&gt; ID1 4.935942 2.853168 3.219038 5.384813 3.320883317</span></span>
<span><span class="co">#&gt; ID2 1.879166 2.699689 3.952487 2.927166 1.934283315</span></span>
<span><span class="co">#&gt; ID3 3.311818 1.078262 1.153573 4.690962 9.238518404</span></span>
<span><span class="co">#&gt; ID4 6.558106 4.683173 1.688233 3.448903 0.007576317</span></span>
<span><span class="co">#&gt; ID5 2.899969 4.930614 2.243714 5.788151 2.843904388</span></span></code></pre></div>
<p>The result is a data frame of VMR IDs as columns and individual IDs
as rows.</p>
</div>
<div class="section level3">
<h3 id="identify-cis-snps">Identify <em>cis</em> SNPs<a class="anchor" aria-label="anchor" href="#identify-cis-snps"></a>
</h3>
<p>After identifying the VMRs, we recommend to use only SNPs in
<em>cis</em> of each VMR, since genetic variants that associate with
DNAme changes tend to be more abundant in the surroundings of the
corresponding DNAme site (McClay <em>et al.</em>, 2015). Also, the
effect sizes of mQTLs (genetic variants associated with DNAme changes)
are stronger in <em>cis</em> SNPs compared to <em>trans</em> SNPs. Then,
by restricting the analysis to <em>cis</em> SNPs, we greatly reduce the
number of variables while keeping most of the important ones.</p>
<p>There is not a clear consensus on how close a SNP has to be from a
DNAme site to be considered <em>cis</em> - the distance threshold tend
to go from few kb to 1 megabase. We recommend to use a 1 Mb window to
cast a wide net to catch potentially relevant SNPs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VMRs_df</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/findCisSNPs.html">findCisSNPs</a></span><span class="op">(</span>VMRs_df <span class="op">=</span> <span class="va">VMRs_df</span>, </span>
<span>                            genotype_information <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_genotype_information.html">test_genotype_information</a></span>, </span>
<span>                            distance <span class="op">=</span> <span class="fl">1e+06</span><span class="op">)</span></span>
<span><span class="co">#&gt; Important: please make sure that the positions of the VMR data frame and the ones in the genotype information are from the same genome build.</span></span>
<span></span>
<span><span class="co">#Take a look at the result</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">VMRs_df</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Rows: 131</span></span>
<span><span class="co">#&gt; Columns: 10</span></span>
<span><span class="co">#&gt; $ VMR_index          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", …</span></span>
<span><span class="co">#&gt; $ seqnames           <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21,…</span></span>
<span><span class="co">#&gt; $ start              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 10990119, 11109021, 15456605, 15588038, 15955548, 1…</span></span>
<span><span class="co">#&gt; $ end                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 10990903, 11109336, 15456789, 15588830, 15956048, 1…</span></span>
<span><span class="co">#&gt; $ width              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 785, 316, 185, 793, 501, 1331, 888, 977, 573, 61, 5…</span></span>
<span><span class="co">#&gt; $ n_VMPs             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 2, 2, 2, 2, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, …</span></span>
<span><span class="co">#&gt; $ probes             <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> &lt;"cg09872009", "cg05437132"&gt;, &lt;"cg00750806", "cg12…</span></span>
<span><span class="co">#&gt; $ median_correlation <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.6099180, 0.6261681, 0.8595934, 0.6829510, 0.75485…</span></span>
<span><span class="co">#&gt; $ surrounding_SNPs   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 1, 1, 610, 667, 726, 794, 757, 529, 709, 788, 829, …</span></span>
<span><span class="co">#&gt; $ SNP                <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> "21:10873592:G:A", "21:10873592:G:A", &lt;"21:1459574…</span></span></code></pre></div>
<p>We can see that the resulting VMRs_df object is almost exactly the
same, but with two new columns (<em>surrounding_SNPs</em> and
<em>SNP</em>) that contain information about how many SNPs were found in
<em>cis</em> and what are their IDs according to the genotype data that
we have.</p>
<p>It is important to highlight the following characteristics of the
VMRs_df object:</p>
<ul>
<li>The <em>VMR_index</em> column is a character vector. This column
corresponds to the unique identifier of each VMR in our data set. It is
important to <strong>keep it as a character</strong> and be careful with
it not being converted to numeric, which can happen if you save the
VMRs_df object as a table, read it, and use that second object in the
rest of the pipeline.</li>
<li>The columns <em>probes</em> and <em>SNP</em> contain
<strong>lists</strong>. This structure is really important for the rest
of the analysis and columns containing lists will keep appearing in
other function outputs. If you want to know the recommended way to save
and load these objects, please check the <a href="#frequently-asked-questions">Frequently Asked Questions</a>.</li>
</ul>
<p>We can also explore the resulting object through plots such as the
following:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VMRs_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>type <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">n_VMPs</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"non canonical"</span>, <span class="co">#non canonical VMRs have only 1 probe by definition</span></span>
<span>                          <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"canonical"</span><span class="op">)</span>,</span>
<span>         surrounding_SNPs <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span> <span class="va">surrounding_SNPs</span> <span class="op">&gt;</span> <span class="fl">3000</span> <span class="op">~</span> <span class="fl">3000</span>,</span>
<span>                                       <span class="cn">TRUE</span> <span class="op">~</span> <span class="va">surrounding_SNPs</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">surrounding_SNPs</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="st">"type"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="RAMEN_files/figure-html/cissnps-1.png" alt="Disribution of SNPs in cis of VMRs." width="700"><p class="caption">
Disribution of SNPs in cis of VMRs.
</p>
</div>
</div>
<div class="section level3">
<h3 id="conduct-variable-selection-on-genome-and-exposome-variables">Conduct variable selection on genome and exposome variables<a class="anchor" aria-label="anchor" href="#conduct-variable-selection-on-genome-and-exposome-variables"></a>
</h3>
<p>The following stage in the pipeline is to screen the available
variables in our environmental and <em>cis</em> SNPs data sets to
identify the potentially relevant ones. This is achieved with the
<code><a href="../reference/selectVariables.html">RAMEN::selectVariables()</a></code> function. This function uses a
data-driven approach based on LASSO, which is an embedded variable
selection method commonly used in machine learning.</p>
<p>In a nutshell, LASSO penalizes models that are more complex (i.e.,
that contain more variables) in favor of simpler models (i.e. that
contain less variables), but not at the expense of reducing predictive
power. Using LASSO’s variable screening property (i.e., with high
probability, the LASSO estimated model includes the substantial
covariates and drops the redundant ones) this function is intended to
select genotype and environmental variables in each Variable Methylated
Region (VMR) with potential relevance in the presence of the
user-specified concomitant variables (which are known DNAme confounders
such as age, cell type proportion, etc.). For more information about the
method, we encourage the users to read the documentation of the
function, and for further information about LASSO we direct readers to
Bühlmann and Van de Geer, 2011.</p>
<p>Overall, conducting our variable selection strategy reduces the
overall computational time and improves the modeling performance by:</p>
<ul>
<li>Reducing the universe of variables that will be used to fit models
in the following stage (G/E/G+E/GxE model fitting and comparison)</li>
<li>Removing redundant variables, which are highly expected in genetic
and environmental data sets with a high number of variables</li>
<li>Limiting the interactions terms to scenarios where both the G and E
main effects were selected to be potentially relevant, which can be
think of as an interaction variable selection using a weak
hierarchy</li>
<li>Using LASSO, a method with good variable selection performance and
scalability</li>
</ul>
<p>Please make sure that your data has no NAs, since the LASSO
implementation we use in RAMEN does not support missing values. If your
data has missing values, consider <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3668100/" class="external-link">handling</a>
them.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_variables</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/selectVariables.html">selectVariables</a></span><span class="op">(</span></span>
<span>  VMRs_df <span class="op">=</span> <span class="va">VMRs_df</span>,</span>
<span>  genotype_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_genotype_matrix.html">test_genotype_matrix</a></span>,</span>
<span>  environmental_matrix<span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_environmental_matrix.html">test_environmental_matrix</a></span>,</span>
<span>  covariates <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_covariates.html">test_covariates</a></span>,</span>
<span>  summarized_methyl_VMR <span class="op">=</span> <span class="va">summarized_methyl_VMR</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: foreach</span></span>
<span><span class="co">#&gt; Loading required package: rngtools</span></span></code></pre></div>
<p>Since LASSO makes use of Random Number Generation, setting a seed is
highly encouraged for result’s reproducibility using the <em>seed</em>
argument. As a note, setting a seed inside of this function modifies the
seed globally (which is R’s default behavior).</p>
<p>The output of <code><a href="../reference/selectVariables.html">RAMEN::selectVariables()</a></code> is an object with
the VMR index, and the G and E variables selected for each VMR.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">selected_variables</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 131</span></span>
<span><span class="co">#&gt; Columns: 3</span></span>
<span><span class="co">#&gt; $ VMR_index      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"…</span></span>
<span><span class="co">#&gt; $ selected_genot <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> "21:10873592:G:A", "21:10873592:G:A", &lt;"21:15450190:C:…</span></span>
<span><span class="co">#&gt; $ selected_env   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> &lt;"E43", "E3", "E5", "E7", "E24", "E25", "E28", "E35", …</span></span></code></pre></div>
<p>We can see how using <code><a href="../reference/selectVariables.html">RAMEN::selectVariables()</a></code> reduces
the number of variables (originally 100 environmental variables and 785
SNPs per VMR on average as seen in Figure @ref(fig:cissnps)).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_variables</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">VMRs_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>              <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">VMR_index</span>,<span class="va">n_VMPs</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            by <span class="op">=</span> <span class="st">"VMR_index"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/transmute.html" class="external-link">transmute</a></span><span class="op">(</span>VMR_index <span class="op">=</span> <span class="va">VMR_index</span>,</span>
<span>            VMR_type <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">n_VMPs</span> <span class="op">&gt;</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"canonical"</span>,</span>
<span>                          <span class="va">n_VMPs</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"non canonical"</span><span class="op">)</span>, </span>
<span>            genotype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">selected_genot</span><span class="op">)</span>,</span>
<span>            environment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">selected_env</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">VMR_index</span>, <span class="va">VMR_type</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">name</span>,</span>
<span>         variables <span class="op">=</span> <span class="va">value</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">VMR_type</span>, y <span class="op">=</span> <span class="va">variables</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html" class="external-link">geom_violin</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>width<span class="op">=</span><span class="fl">0.1</span>, outlier.shape<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">group</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Selected variables"</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="RAMEN_files/figure-html/selectedvars-1.png" alt="Number of G and E selected variables." width="700"><p class="caption">
Number of G and E selected variables.
</p>
</div>
<p>It is also expected in real data to have VMRs where no SNP and/or no
environmental variables were selected, since not all the DNAme sites in
the genome are expected to show an association with the genetic
variation or environmental exposures data sets that are captured in a
study. The proportion of VMRs under these scenarios will depend on the
data sets.</p>
<div class="section level4">
<h4 id="authors-note-about-variables-interpretation">Author’s note about variables interpretation<a class="anchor" aria-label="anchor" href="#authors-note-about-variables-interpretation"></a>
</h4>
<p>LASSO variable selection is not consistent when there is
multicollinearity in the data (i.e., correlation between variables),
which is expected due to the high amount of G and E variables that are
present in studies of this kind. This means that if you were to run
LASSO several times, and two variables were to be highly correlated, the
method would select one and drop the other one at random. This is not a
problem with the pipeline because the main conclusion per VMR is whether
the DNAme is better explained by G and/or E components. As an example,
if a VMR is better explained by SNP1 and SNP2, which are both highly
correlated one with the other, LASSO will randomly pick SNP1 OR SNP2
(because they are relevant but they provide redundant information); if
we were to fit a model with SNP1 or SNP2 in the following stage, the
winning model would still be G. In other words, the main goal of the
pipeline is to know whether the VMR’s DNAme is better explained by G
and/or E. The user is therefore warned to <strong>be cautious not to
over-interpret the individual selected variables</strong>. Selected
variables might be used as hypothesis generators of associations,
keeping in mind that the selected variable might be representing other
variables in the data set that provide similar information.</p>
</div>
</div>
<div class="section level3">
<h3 id="identify-the-best-explanatory-model-gegegxe-per-vmr">Identify the best explanatory model (G/E/G+E/GxE) per VMR<a class="anchor" aria-label="anchor" href="#identify-the-best-explanatory-model-gegegxe-per-vmr"></a>
</h3>
<p>Now that we have selected the list of potentially relevant G and E
variables, we will fit the models mentioned in Table
@ref(tab:modelstable) using the <code><a href="../reference/lmGE.html">RAMEN::lmGE()</a></code> function.
This function fits, for each VMR, G and E models with all of the
variables selected, as well as all their possible pairwise combinations
of G+E and GxE.</p>
<p>After fitting this model, the best model per group (group = G, E, G+E
or GxE) is selected using Akaike Information Criterion (AIC) or Bayesian
Information Criterion (BIC). We recommend using AIC because BIC assumes
that the true model is in the set of compared models. Since this
function fits models with individual variables, and we assume that DNAme
variability is more likely to be influenced by more than one single
SNP/environmental exposure at a time, we hypothesize that in most cases,
the true model will not be in the set of compared models. Also, AIC
excels in situations where all models in the model space are
“incorrect”, and AIC is preferentially used in cases where the true
underlying function is unknown and our selected model could belong to a
very large class of functions where the relationship could be pretty
complex. It is worth mentioning however that, both metrics tend to pick
the same model in a large number of scenarios. We suggest the users to
read Arijit Chakrabarti &amp; Jayanta K. Ghosh, 2011 for further
information about the difference between these metrics. After selecting
the best model per group (G,E,G+E pr GxE), the model with the lowest AIC
or BIC will be declared as the winning model.</p>
<p>Additionally, <code><a href="../reference/lmGE.html">RAMEN::lmGE()</a></code> conducts a variance
decomposition analysis, so that the relative R2 contribution of each of
the variables of interest (G, E and GxE) is reported. This decomposition
is done using the <em><a href="https://CRAN.R-project.org/package=relaimpo" class="external-link">relaimpo</a></em> R
package, using the Lindeman, Merenda and Gold (lmg) method, which is
based on the heuristic approach of averaging the relative R contribution
of each variable over all input orders in the linear model.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmge_res</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/lmGE.html">lmGE</a></span><span class="op">(</span></span>
<span>  selected_variables <span class="op">=</span> <span class="va">selected_variables</span>,</span>
<span>  summarized_methyl_VMR <span class="op">=</span> <span class="va">summarized_methyl_VMR</span>,</span>
<span>  genotype_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_genotype_matrix.html">test_genotype_matrix</a></span>,</span>
<span>  environmental_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_environmental_matrix.html">test_environmental_matrix</a></span>,</span>
<span>  covariates <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_covariates.html">test_covariates</a></span>,</span>
<span>  model_selection <span class="op">=</span> <span class="st">"AIC"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check the output</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">lmge_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 128</span></span>
<span><span class="co">#&gt; Columns: 13</span></span>
<span><span class="co">#&gt; $ VMR_index       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11…</span></span>
<span><span class="co">#&gt; $ model_group     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "G+E", "G+E", "GxE", "GxE", "G+E", "GxE", "G+E", "GxE"…</span></span>
<span><span class="co">#&gt; $ variables       <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> &lt;"21:10873592:G:A", "E43"&gt;, &lt;"21:10873592:G:A", "E43"…</span></span>
<span><span class="co">#&gt; $ tot_r_squared   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.5527507, 0.5115764, 0.6075657, 0.6794845, 0.7680461,…</span></span>
<span><span class="co">#&gt; $ g_r_squared     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.1996156, 0.2095645, 0.3064024, 0.2738786, 0.3981840,…</span></span>
<span><span class="co">#&gt; $ e_r_squared     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.34206198, 0.27143559, 0.15293109, 0.20266291, 0.2769…</span></span>
<span><span class="co">#&gt; $ gxe_r_squared   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> NA, NA, 0.1330680, 0.1126333, NA, 0.1432113, NA, 0.108…</span></span>
<span><span class="co">#&gt; $ AIC             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 144.6841, 148.7250, 145.6987, 135.7819, 129.9252, 123.…</span></span>
<span><span class="co">#&gt; $ second_winner   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "GxE", "GxE", "G+E", "G+E", "GxE", "G+E", "GxE", "G+E"…</span></span>
<span><span class="co">#&gt; $ delta_aic       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 1.04970268, 1.34557464, 1.01022314, 3.88212651, 0.0761…</span></span>
<span><span class="co">#&gt; $ delta_r_squared <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> -0.013945293, -0.010539189, 0.041420496, 0.069427874, …</span></span>
<span><span class="co">#&gt; $ basal_AIC       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 164.4893, 165.2905, 167.3019, 161.0771, 166.8374, 159.…</span></span>
<span><span class="co">#&gt; $ basal_rsquared  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.011073059, 0.030576374, 0.015164242, 0.090309667, 0.…</span></span></code></pre></div>
<p>The output of <code><a href="../reference/lmGE.html">RAMEN::lmGE()</a></code> is a data frame with the
following 13 columns:</p>
<ul>
<li>
<em>VMR_index</em>: The index of the respective VMR</li>
<li>
<em>model_group</em>: The selected winning model (G, E, G+E or
GxE)</li>
<li>
<em>variables</em>: The variable(s) that are present in the winning
model (excluding the covariates, which are included in all the
models)</li>
<li>
<em>tot_r_squared</em>: total R squared of the winning model</li>
<li>
<em>g_r_squared</em>: Estimated R2 allocated to the G component in
the winning model, if applicable</li>
<li>
<em>e_r_squared</em>: Estimated R2 allocated to the E in the winning
model, if applicable.</li>
<li>
<em>gxe_r_squared</em>: Estimated R2 allocated to the interaction in
the winning model (GxE), if applicable.</li>
<li>
<em>AIC/BIC</em>: AIC or BIC metric from the best model in each VMR
(depending on the option specified in the argument
model_selection).</li>
<li>
<em>second_winner</em>: The second group that possesses the next
best model after the winning one (i.e., G, E, G+E or GxE). This column
may have NA if the variables in selected_variables correspond only to
one group (G or E), so that there is no other model groups to compare
to.</li>
<li>
<em>delta_aic/delta_bic</em>: The difference of AIC or BIC
(depending on the option specified in the argument model_selection) of
the winning model and the best model from the second_winner group (i.e.,
G, E, G+E or GxE). This column may have NA if the variables in
selected_variables correspond only to one group (G or E), so that there
is no other groups to compare to.</li>
<li>
<em>delta_r_squared</em>: The R2 of the winning model - R2 of the
second winner model. This column may have NA if the variables in
selected_variables correspond only to one group (G or E), so that there
is no other groups to compare to.</li>
<li>
<em>basal_AIC/basal_BIC</em>: AIC or BIC of the basal model (i.e.,
model fitted only with the concomitant variables specified in the
covariates argument)</li>
<li>
<em>basal_rsquared</em>: The R2 of the basal model (i.e., model
fitted only with the concomitant variables specified in the covariates
argument)</li>
</ul>
<div class="section level4">
<h4 id="remove-poor-performing-winning-models">Remove poor performing winning models<a class="anchor" aria-label="anchor" href="#remove-poor-performing-winning-models"></a>
</h4>
<p>The core pipeline from the RAMEN package identifies the best
explanatory model per VMR. However, despite these models being winners
in comparison to models including any other G/E variable(s) in the
dataset, some winning models might perform no better than what we would
expect by chance. Therefore, The last step of the pipeline after
identifying the winning models per VMR is to compute a null distribution
to remove the winning models that are very likely to be winners by
chance. To do so, we use <code><a href="../reference/nullDistGE.html">RAMEN::nullDistGE()</a></code>.</p>
<p>The goal of <code><a href="../reference/nullDistGE.html">RAMEN::nullDistGE()</a></code> is to create a
distribution of increase in R2 after including the SNP/E/SNP*E variables
under the null hypothesis of G and E having no associations with DNAme.
The null distribution is obtained through shuffling the G and E
variables in a given dataset and conducting the variable selection and
G/E model selection. That way, we can simulate how much additional
variance would be explained by the models defined as winners by the
RAMEN methodology in a scenario where the G and E associations with
DNAme are randomized. This distribution can be then used to filter out
winning models in the non-shuffled dataset that do not add more to the
explained variance of the basal model than what randomized data do.</p>
<p>For clarification, please note that in this vignette when we refer to
SNP*E, we are referring to the interaction term that is present in the
the interaction model (i.e. interaction variable in the GxE model).</p>
<p>Under the assumption that after adjusting for the concomitant
variables all VMRs across the genome follow the same behavior regarding
an increment of explained variance with randomized G and E data, we can
pool the delta R squared values from all VMRs to create a null
distribution taking advantage of the high number of VMRs in the dataset.
This assumption decreases significantly the number of permutations
required to create a null distribution and reduces the computational
time. For further information please read the RAMEN paper (in
preparation).</p>
<p><code><a href="../reference/nullDistGE.html">RAMEN::nullDistGE()</a></code> simulates the delta R squared
distribution under the null hypothesis of G and E having no association
with DNA methylation (DNAme) variability through a permutation analysis.
To do so, this function shuffles the G and E variables in the dataset,
which is followed by a the variable selection and modelling steps with
selectVariables() and lmGE().These steps are repeated several times as
indicated in the <em>permutations</em> parameter. In other words, by
using shuffled G and E data, we simulate the increase of R2 that would
be observed in random data using the RAMEN methodology.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute the null distribution </span></span>
<span><span class="va">null_dist</span> <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="fu"><a href="../reference/nullDistGE.html">nullDistGE</a></span><span class="op">(</span></span>
<span>  VMRs_df <span class="op">=</span> <span class="va">VMRs_df</span>,</span>
<span>  genotype_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_genotype_matrix.html">test_genotype_matrix</a></span>,</span>
<span>  environmental_matrix <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_environmental_matrix.html">test_environmental_matrix</a></span>,</span>
<span>  summarized_methyl_VMR <span class="op">=</span> <span class="va">summarized_methyl_VMR</span>,</span>
<span>  permutations <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  covariates <span class="op">=</span> <span class="fu">RAMEN</span><span class="fu">::</span><span class="va"><a href="../reference/test_covariates.html">test_covariates</a></span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  model_selection <span class="op">=</span> <span class="st">"AIC"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Take a look at the object </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">null_dist</span><span class="op">)</span></span>
<span><span class="co">#&gt;   VMR_index tot_r_squared model_group R2_difference AIC_difference permutation</span></span>
<span><span class="co">#&gt; 1         1     0.2606017           E     0.2495286       157.7547           1</span></span>
<span><span class="co">#&gt; 2         2     0.2976822           E     0.2671059       157.5905           1</span></span>
<span><span class="co">#&gt; 3         3     0.6086939         G+E     0.5935297       143.5972           1</span></span>
<span><span class="co">#&gt; 4         4     0.3777878           E     0.2874781       151.5921           1</span></span>
<span><span class="co">#&gt; 5         6     0.6133220         G+E     0.5269145       137.8163           1</span></span>
<span><span class="co">#&gt; 6         7     0.3451632           E     0.2497069       155.1677           1</span></span></code></pre></div>
<p>The output is a data frame where the most useful column for our
purpose is <em>R2_difference</em>, which corresponds to the increase in
R squared obtained by including the G/E variable(s) from the winning
model (i.e., the R squared difference between the winning model and the
model only with the concomitant variables specified in covariates;
tot_r_squared - basal_rsquared in the lmGE output)</p>
<p>We recommend to use two different thresholds for the winning models
depending of whether they are marginal (G or E) or joint models (G+E or
GxE). The reason for this is that they have different R2_difference
distributions. E and G models have a lower mean R2_difference because
they have a single shuffled term in the model (SNP or E). In comparison,
joint models have a higher mean R2_difference because they have two or
three shuffled terms (SNP, E and SNP*E), which just by chance increases
their probability of having a higher R2_difference.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># See the distribution of R2_difference across different winning models</span></span>
<span><span class="va">null_dist</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">R2_difference</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="st">"model_group"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre></div>
<div class="figure">
<img src="RAMEN_files/figure-html/unnamed-chunk-13-1.png" alt="R2 difference (winner - basal) in a suffled data set." width="700"><p class="caption">
R2 difference (winner - basal) in a suffled data set.
</p>
</div>
<p>We suggest using the 95th percentile of those distributions as a
threshold to remove bad performing winning models found in our observed
data.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a cutoff of the 95th percentile of the null distribution for single and joint models</span></span>
<span><span class="va">cutoff_single</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">null_dist</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>                           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">model_group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G"</span>,<span class="st">"E"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>                           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">R2_difference</span><span class="op">)</span>, </span>
<span>                         <span class="fl">0.95</span><span class="op">)</span></span>
<span><span class="va">cutoff_joint</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">null_dist</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>                           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">model_group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G+E"</span>,<span class="st">"GxE"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>                           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">R2_difference</span><span class="op">)</span>, </span>
<span>                         <span class="fl">0.95</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Filter out bad performing winning models.</span></span>
<span><span class="va">filtered_res</span> <span class="op">=</span> <span class="va">lmge_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>r2_difference_basal <span class="op">=</span> <span class="va">tot_r_squared</span> <span class="op">-</span> <span class="va">basal_rsquared</span>,</span>
<span>         pass_cutoff_threshold <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">model_group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G"</span>, <span class="st">"E"</span><span class="op">)</span> <span class="op">~</span> <span class="va">r2_difference_basal</span> <span class="op">&gt;</span> <span class="va">cutoff_single</span>,</span>
<span>                                           <span class="va">model_group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G+E"</span>, <span class="st">"GxE"</span><span class="op">)</span> <span class="op">~</span> <span class="va">r2_difference_basal</span> <span class="op">&gt;</span> <span class="va">cutoff_joint</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pass_cutoff_threshold</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co">#Filter based on the cutoff threshold</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">pass_cutoff_threshold</span><span class="op">)</span> <span class="co">#Drop temporary column</span></span>
<span></span>
<span><span class="co">#Check the final results</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">filtered_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 7</span></span>
<span><span class="co">#&gt; Columns: 14</span></span>
<span><span class="co">#&gt; $ VMR_index           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "5", "6", "28", "39", "96", "98", "106"</span></span>
<span><span class="co">#&gt; $ model_group         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "G+E", "GxE", "E", "GxE", "G", "G", "GxE"</span></span>
<span><span class="co">#&gt; $ variables           <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> &lt;"21:15671097:T:A", "E43"&gt;, &lt;"21:15671097:T:A", "E…</span></span>
<span><span class="co">#&gt; $ tot_r_squared       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.7680461, 0.7769276, 0.5555035, 0.7097936, 0.6431…</span></span>
<span><span class="co">#&gt; $ g_r_squared         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.39818402, 0.38220942, NA, 0.09101486, 0.6398122…</span></span>
<span><span class="co">#&gt; $ e_r_squared         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.2769884, 0.1650993, 0.5244324, 0.2406247, NA, NA…</span></span>
<span><span class="co">#&gt; $ gxe_r_squared       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> NA, 0.1432113, NA, 0.3522698, NA, NA, 0.1515714</span></span>
<span><span class="co">#&gt; $ AIC                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 129.9252, 123.3999, 145.1528, 147.3162, 140.8549, …</span></span>
<span><span class="co">#&gt; $ second_winner       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "GxE", "G+E", NA, "G+E", "G+E", NA, "G+E"</span></span>
<span><span class="co">#&gt; $ delta_aic           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.07612103, 3.90964213, NA, 3.27939276, 0.34990027…</span></span>
<span><span class="co">#&gt; $ delta_r_squared     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> -0.01440811, 0.04856945, NA, 0.05583976, -0.019097…</span></span>
<span><span class="co">#&gt; $ basal_AIC           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 166.8374, 159.6965, 166.5303, 177.6443, 169.6680, …</span></span>
<span><span class="co">#&gt; $ basal_rsquared      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.092873702, 0.086407501, 0.031071178, 0.025884145…</span></span>
<span><span class="co">#&gt; $ r2_difference_basal <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.6751724, 0.6905201, 0.5244324, 0.6839094, 0.6398…</span></span></code></pre></div>
<p>We can see that the final data set in this example dropped almost all
of the VMRs we had. This is something expected since we are working with
toy data coming from random sampling.</p>
<p>We recommend the users of the package to include the number of VMRs
where we could not find a conclusive best model in the final results
report (either because no variables were selected with
<code><a href="../reference/selectVariables.html">RAMEN::selectVariables()</a></code> or because they did not pass the
R2_difference threshold obtained with
<code><a href="../reference/nullDistGE.html">RAMEN::nullDistGE()</a></code>).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Include the VMRs with no winning model in the final results object</span></span>
<span><span class="va">final_res</span> <span class="op">=</span> <span class="va">VMRs_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">filtered_res</span>,</span>
<span>                    by <span class="op">=</span> <span class="st">"VMR_index"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot final results</span></span>
<span><span class="va">final_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">model_group</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>count <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">model_group</span>, y <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Best explanatory model"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"VMRs"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<div class="figure">
<img src="RAMEN_files/figure-html/finalresults-1.png" alt="Variable Methylated Regions best explanatory models" width="700"><p class="caption">
Variable Methylated Regions best explanatory models
</p>
</div>
<p>So, we can see that for this toy example, we got the following
results:</p>
<ul>
<li>VMRs better explained by a G model: 2<br>
</li>
<li>VMRs better explained by a E model: 1</li>
<li>VMRs better explained by a G+E model: 1</li>
<li>VMRs better explained by a GxE model: 3</li>
<li>VMRs with no conclusive explanatory model: 124</li>
</ul>
</div>
<div class="section level4">
<h4 id="authors-note-about-model-interpretation">Author’s note about model interpretation<a class="anchor" aria-label="anchor" href="#authors-note-about-model-interpretation"></a>
</h4>
<p>For model simplicity, each winning model have a single E and/or G
variable (and its interaction term when applicable). That means that in
a scenario where a given VMR is under the influence of 2 or more Es or
Gs, only the one that better explains the VMR’s DNAme alone will be
selected. In other words, if a VMR in reality is influenced by
e.g. folate intake and smoking, and we have information about both
environmental exposures, the best model (E) will have only folate intake
(in case that is the variable that better explains DNAme variability in
that region alone). So, interpreting this as the VMR not being
potentially under the influence of smoking might not be correct. We
recommend the user to check the total R2 of the winning model to explore
the remaining variance that is not explained by the winning model.</p>
<p>We also stress that <strong>interpretation of individual variables
should be done with caution and used only as exploration and research
hypothesis generation</strong>. Please see <a href="#authors-note-about-variables-interpretation">Author’s note about
variables interpretation</a>) where we advice against
over-interpretation of the selected variables; the same logic applies to
the variables present in the winning models.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="variations-to-the-standard-workflow">Variations to the standard workflow<a class="anchor" aria-label="anchor" href="#variations-to-the-standard-workflow"></a>
</h2>
<p>Besides using RAMEN for completing the analysis mentioned above, the
outputs of the package’s function can help users in other DNAme
analyses, such as:</p>
<ul>
<li>Reduction of tests prior to an EWAS or differential methylation
analysis with <code><a href="../reference/findVMRs.html">RAMEN::findVMRs()</a></code>(i.e., conducting the
analysis on identified VMRs which 1) reduces redundant tests by grouping
nearby correlated CpGs, and 2) avoids tests in non-variant regions).
This can help to reduce the multiple hypothesis testing burden.</li>
<li>Summarize a DNAme region of interest with
<code><a href="../reference/summarizeVMRs.html">RAMEN::summarizeVMRs()</a></code>
</li>
<li>Easily conduct variable selection in high-dimensional data sets to
identify potentially relevant variables from one or two independent data
sets with <code><a href="../reference/selectVariables.html">RAMEN::selectVariables()</a></code>.</li>
<li>Fit additive and interaction models given two sets of variables of
interest (not limited to G and E) and select the best explanatory model
for DNAme data with <code><a href="../reference/selectVariables.html">RAMEN::selectVariables()</a></code> and
<code><a href="../reference/lmGE.html">RAMEN::lmGE()</a></code> (e.g. exploring the interaction between two
environmental dimensions and their contribution to DNAme
variability).</li>
<li>Quickly identify SNPs in <em>cis</em> of CpG probes with
<code><a href="../reference/findCisSNPs.html">RAMEN::findCisSNPs()</a></code> (e.g. for cis mQTL analyses)</li>
<li>Get the median correlation of probes in regions of interest (with
<code><a href="../reference/medCorVMR.html">RAMEN::medCorVMR()</a></code>).</li>
</ul>
</div>
<div class="section level2">
<h2 id="frequently-asked-questions">Frequently Asked Questions<a class="anchor" aria-label="anchor" href="#frequently-asked-questions"></a>
</h2>
<p><strong>How can I save the RAMEN data frames that have columns with
lists as observations?</strong></p>
<p>Saving the data frames produced by RAMEN might seem difficult because
it has lists as observations in several columns, which is not supported
by some read/write functions. We suggest two options:</p>
<ol style="list-style-type: decimal">
<li>Use <code>fwrite()</code> and <code>fread()</code> from the <em><a href="https://CRAN.R-project.org/package=data.table" class="external-link">data.table</a></em>
package (recommended because of time and storage space).</li>
</ol>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example for saving and reading selected_variables object</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu">fwrite</span><span class="op">(</span><span class="va">selected_variables</span>, file <span class="op">=</span> <span class="st">"path/selected_variables.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read the csv file and make lists the elements in the required columns</span></span>
<span><span class="va">selected_variables</span> <span class="op">=</span> <span class="fu">fread</span><span class="op">(</span><span class="st">"path/selected_variables.csv"</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>selected_genot <span class="op">=</span> <span class="fu">str_split</span><span class="op">(</span><span class="va">selected_genot</span>, pattern <span class="op">=</span> <span class="st">"\\|"</span><span class="op">)</span>, <span class="co"># fwrite saves lists as strings separated by |</span></span>
<span>         selected_env <span class="op">=</span><span class="fu">str_split</span><span class="op">(</span><span class="va">selected_env</span>, pattern <span class="op">=</span> <span class="st">"\\|"</span><span class="op">)</span>,</span>
<span>         VMR_index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">VMR_index</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Save files as .rds</li>
</ol>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example for saving the selected_variables object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">selected_variables</span>, file <span class="op">=</span> <span class="st">"path/selected_variables.Rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Load the object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"path/selected_variables.Rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] doRNG_1.8.6      rngtools_1.5.2   foreach_1.5.2    tidyr_1.3.1     </span></span>
<span><span class="co">#&gt; [5] ggplot2_3.5.1    dplyr_1.1.4      RAMEN_1.0.0      knitr_1.48      </span></span>
<span><span class="co">#&gt; [9] BiocStyle_2.32.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gtable_0.3.5            shape_1.4.6.1           xfun_0.48              </span></span>
<span><span class="co">#&gt;  [4] bslib_0.8.0             lattice_0.22-6          vctrs_0.6.5            </span></span>
<span><span class="co">#&gt;  [7] tools_4.4.1             generics_0.1.3          stats4_4.4.1           </span></span>
<span><span class="co">#&gt; [10] parallel_4.4.1          tibble_3.2.1            fansi_1.0.6            </span></span>
<span><span class="co">#&gt; [13] highr_0.11              pkgconfig_2.0.3         Matrix_1.7-0           </span></span>
<span><span class="co">#&gt; [16] desc_1.4.3              S4Vectors_0.42.1        lifecycle_1.0.4        </span></span>
<span><span class="co">#&gt; [19] GenomeInfoDbData_1.2.12 compiler_4.4.1          farver_2.1.2           </span></span>
<span><span class="co">#&gt; [22] stringr_1.5.1           textshaping_0.4.0       munsell_0.5.1          </span></span>
<span><span class="co">#&gt; [25] mitools_2.4             codetools_0.2-20        survey_4.4-2           </span></span>
<span><span class="co">#&gt; [28] GenomeInfoDb_1.40.1     htmltools_0.5.8.1       sass_0.4.9             </span></span>
<span><span class="co">#&gt; [31] yaml_2.3.10             glmnet_4.1-8            pillar_1.9.0           </span></span>
<span><span class="co">#&gt; [34] pkgdown_2.1.1           jquerylib_0.1.4         MASS_7.3-60.2          </span></span>
<span><span class="co">#&gt; [37] cachem_1.1.0            iterators_1.0.14        boot_1.3-30            </span></span>
<span><span class="co">#&gt; [40] tidyselect_1.2.1        digest_0.6.37           stringi_1.8.4          </span></span>
<span><span class="co">#&gt; [43] purrr_1.0.2             bookdown_0.40           labeling_0.4.3         </span></span>
<span><span class="co">#&gt; [46] splines_4.4.1           fastmap_1.2.0           grid_4.4.1             </span></span>
<span><span class="co">#&gt; [49] colorspace_2.1-1        cli_3.6.3               magrittr_2.0.3         </span></span>
<span><span class="co">#&gt; [52] survival_3.6-4          utf8_1.2.4              corpcor_1.6.10         </span></span>
<span><span class="co">#&gt; [55] withr_3.0.1             scales_1.3.0            UCSC.utils_1.0.0       </span></span>
<span><span class="co">#&gt; [58] rmarkdown_2.28          XVector_0.44.0          httr_1.4.7             </span></span>
<span><span class="co">#&gt; [61] ragg_1.3.3              evaluate_1.0.0          GenomicRanges_1.56.1   </span></span>
<span><span class="co">#&gt; [64] IRanges_2.38.1          rlang_1.1.4             Rcpp_1.0.13            </span></span>
<span><span class="co">#&gt; [67] DBI_1.2.3               glue_1.8.0              BiocManager_1.30.25    </span></span>
<span><span class="co">#&gt; [70] BiocGenerics_0.50.0     jsonlite_1.8.9          R6_2.5.1               </span></span>
<span><span class="co">#&gt; [73] relaimpo_2.2-7          systemfonts_1.1.0       fs_1.6.4               </span></span>
<span><span class="co">#&gt; [76] zlibbioc_1.50.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Erick I. Navarro-Delgado.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
